{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d9f808",
   "metadata": {},
   "source": [
    "# Preprocess Gene Annotations\n",
    "\n",
    "This notebook creates a table of gene annotations by:\n",
    "1. Querying Biomart for all Ensembl IDs in the database\n",
    "2. Querying MyGene for annotation about those IDs\n",
    "3. Querying Ensembl for the most recent Ensembl release for each ID\n",
    "4. Building a permalink to the Ensembl archive page for each ID\n",
    "\n",
    "This gene annotation table is read in by `agoradataprocessing/process.py` to be used in the `gene_info` transformation. \n",
    "\n",
    "***Note:*** *This notebook is exploratory and should eventually be converted to a Python script that is run through an automated process.*\n",
    "\n",
    "## Installation requirements\n",
    "\n",
    "#### Linux / Windows / Mac\n",
    "\n",
    "Install R: https://cran.r-project.org/\n",
    "\n",
    "Install Python and agora-data-tools following the instructions in this repository's README. This notebook assumes it is being run from the same `pipenv` virtual environment as agora-data-tools. \n",
    "\n",
    "Then install the following packages using `pip`:\n",
    "```\n",
    "pip install rpy2 mygene\n",
    "```\n",
    "\n",
    "#### Note for Macs with M1 chips (2020 and newer)\n",
    "\n",
    "Install as above, but make sure that your R installation is the arm64 version (R-4.X.X-arm64.pkg) so that the architecture matches what pip is using. \n",
    "You may also need to install an older version of `rpy2` on the Mac:\n",
    "```\n",
    "pip install rpy2==3.5.12\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2dc92c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocManager\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.0 (2023-04-21)\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version '3.17' is out-of-date; the current release version '3.18'\n",
      "  is available with R version '4.3'; see https://bioconductor.org/install\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: biomaRt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from rpy2.robjects import r\n",
    "import pandas as pd\n",
    "import mygene\n",
    "import numpy as np\n",
    "import requests\n",
    "import agoradatatools.etl.utils as utils\n",
    "import agoradatatools.etl.extract as extract\n",
    "import preprocessing_utils\n",
    "\n",
    "r('if (!require(\"BiocManager\", character.only = TRUE)) { install.packages(\"BiocManager\") }')\n",
    "r('if (!require(\"biomaRt\")) { BiocManager::install(\"biomaRt\") }')\n",
    "\n",
    "r.library('biomaRt')\n",
    "\n",
    "biomart_filename = '../../output/biomart_ensg_list.txt'\n",
    "archive_filename = '../../output/ensembl_archive_list.csv'\n",
    "config_filename = '../../../../config.yaml'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec36f54c",
   "metadata": {},
   "source": [
    "# Part 1: Get gene annotation data\n",
    "\n",
    "## Query Biomart for a list of all Ensembl IDs in the database of human genes. \n",
    "\n",
    "Here we use the R library `biomaRt`. There is no canonical Python library with the features we need for this script. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d1e115d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Ensembl site unresponsive, trying asia mirror\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       ensembl_gene_id chromosome_name hgnc_symbol\n",
      "0      ENSG00000210049              MT       MT-TF\n",
      "1      ENSG00000211459              MT     MT-RNR1\n",
      "2      ENSG00000210077              MT       MT-TV\n",
      "3      ENSG00000210082              MT     MT-RNR2\n",
      "4      ENSG00000209082              MT      MT-TL1\n",
      "...                ...             ...         ...\n",
      "70707  ENSG00000288629               1            \n",
      "70708  ENSG00000288678               1            \n",
      "70709  ENSG00000290825               1            \n",
      "70710  ENSG00000227232               1      WASH7P\n",
      "70711  ENSG00000290826               1            \n",
      "\n",
      "[70712 rows x 3 columns]\n",
      "70712 genes found.\n"
     ]
    }
   ],
   "source": [
    "# Sometimes Biomart doesn't respond and the command needs to be sent again. Try up to 5 times.\n",
    "for T in range(5):\n",
    "    try:\n",
    "        mart = r.useEnsembl(biomart = 'ensembl', dataset = 'hsapiens_gene_ensembl')\n",
    "        ensembl_ids = r.getBM(attributes = r.c('ensembl_gene_id', 'chromosome_name', 'hgnc_symbol'), mart = mart, useCache = False)\n",
    "        \n",
    "    except:\n",
    "        print('Trying again...')\n",
    "        ensembl_ids = None\n",
    "    else: \n",
    "        break\n",
    "\n",
    "if ensembl_ids is None or ensembl_ids.nrow == 0:\n",
    "    print('Biomart was unresponsive after 5 attempts. Try again later.')\n",
    "\n",
    "else:\n",
    "    # Convert the ensembl_gene_id column from R object to a python list\n",
    "    ensembl_ids_df = pd.DataFrame({\"ensembl_gene_id\": list(ensembl_ids.rx2('ensembl_gene_id')), \n",
    "                                   \"chromosome_name\": list(ensembl_ids.rx2('chromosome_name')),\n",
    "                                   \"hgnc_symbol\": list(ensembl_ids.rx2('hgnc_symbol'))})\n",
    "    print(ensembl_ids_df)\n",
    "    print(str(ensembl_ids_df.shape[0]) + \" genes found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ddd4e52",
   "metadata": {},
   "source": [
    "## Remove HASGs from Biomart results\n",
    "This removes all human alternative sequence genes (HASGs) and patches from the Biomart results. These genes can be identified by their chromosome name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1709e32c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_ids_df = preprocessing_utils.filter_HASGs(df=ensembl_ids_df, chromosome_name_column=\"chromosome_name\")\n",
    "print(str(ensembl_ids_df.shape[0]) + \" genes remaining after HASG filtering.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f494e58f",
   "metadata": {},
   "source": [
    "## Add Ensembl IDs from data sets that will be processed by agora-data-tools\n",
    "\n",
    "Some of these datasets have older/retired Ensembl IDs that no longer exist in the current Ensembl database. Loop through all data sets in the config file to get all Ensembl IDs they use, and add missing ones to the gene list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a3fdbeec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes_biodomains': ('syn44151254.1', 'csv'),\n",
       " 'neuropath_regression_results': ('syn22017882.5', 'csv'),\n",
       " 'proteomics': ('syn18689335.3', 'csv'),\n",
       " 'proteomics_tmt': ('syn35221005.2', 'csv'),\n",
       " 'proteomics_srm': ('syn52579640.4', 'csv'),\n",
       " 'target_exp_validation_harmonized': ('syn24184512.8', 'csv'),\n",
       " 'metabolomics': ('syn26064497.1', 'feather'),\n",
       " 'igap': ('syn12514826.5', 'csv'),\n",
       " 'eqtl': ('syn12514912.3', 'csv'),\n",
       " 'diff_exp_data': ('syn27211942.1', 'tsv'),\n",
       " 'target_list': ('syn12540368.47', 'csv'),\n",
       " 'median_expression': ('syn27211878.2', 'csv'),\n",
       " 'druggability': ('syn13363443.11', 'csv'),\n",
       " 'tep_adi_info': ('syn51942280.2', 'csv'),\n",
       " 'team_info': ('syn12615624.18', 'csv'),\n",
       " 'team_member_info': ('syn12615633.18', 'csv'),\n",
       " 'overall_scores': ('syn25575156.13', 'table'),\n",
       " 'networks': ('syn11685347.1', 'csv')}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = utils._get_config(config_path = config_filename)\n",
    "datasets = config['datasets']\n",
    "\n",
    "files = {}\n",
    "\n",
    "for dataset in datasets:\n",
    "    dataset_name = list(dataset.keys())[0]\n",
    "    \n",
    "    for entity in dataset[dataset_name]['files']:\n",
    "        entity_id = entity['id']\n",
    "        entity_format = entity['format']\n",
    "        entity_name = entity['name']\n",
    "        \n",
    "        # Ignore json files, which are post-processed and not what we're interested in. \n",
    "        # Also ignore \"gene_metadata\" since that's the file we're making here.\n",
    "        if entity_format != 'json' and entity_name != \"gene_metadata\":\n",
    "            files[entity_name] = (entity_id, entity_format)\n",
    "\n",
    "# There are some duplicate synID's in this list but that doesn't really matter\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1a2120",
   "metadata": {},
   "source": [
    "### We should now have a list of all raw data files ingested. Get each one and add its Ensembl IDs to the gene list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9843689d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.0.0) is available. Your version (3.1.1) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.0.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:synapseclient_default:Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target_exp_validation_harmonized has an n/A Ensembl ID\n"
     ]
    }
   ],
   "source": [
    "syn = utils._login_to_synapse(token = None) # Assumes you have already logged in with a valid token\n",
    "\n",
    "# The various column names used to store Ensembl IDs in the files\n",
    "col_names = ['ENSG', 'ensembl_gene_id', 'GeneID']\n",
    "file_ensembl_list = []\n",
    "\n",
    "for file in files.keys():\n",
    "    df = extract.get_entity_as_df(syn_id=files[file][0],\n",
    "                                  source=files[file][1],\n",
    "                                  syn=syn)\n",
    "    file_ensembl_ids = None\n",
    "    \n",
    "    for C in col_names:\n",
    "        if C in df.columns:\n",
    "            file_ensembl_ids = df[C]\n",
    "    \n",
    "    # networks file is a special case\n",
    "    if file == \"networks\":\n",
    "        file_ensembl_ids = pd.melt(df[['geneA_ensembl_gene_id','geneB_ensembl_gene_id']])['value']\n",
    "        \n",
    "    if file_ensembl_ids is not None:\n",
    "        file_ensembl_list = file_ensembl_list + file_ensembl_ids.tolist()\n",
    "        if \"n/A\" in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an n/A Ensembl ID\")\n",
    "        if np.NaN in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an NaN Ensembl ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1303e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1165 genes from the data files are missing from Biomart results and will be added.\n",
      "71877\n"
     ]
    }
   ],
   "source": [
    "file_ensembl_list = list(set(file_ensembl_list))\n",
    "file_ensembl_list.remove(\"n/A\") # Necessary because one data set has an Ensembl ID set to \"n/A\"\n",
    "\n",
    "# Add Ensembl IDs that are in the files but not in the biomart result\n",
    "missing = set(file_ensembl_list) - set(ensembl_ids_df['ensembl_gene_id'])\n",
    "print(str(len(missing)) + \" genes from the data files are missing from Biomart results and will be added.\")\n",
    "\n",
    "missing_df = pd.DataFrame({'ensembl_gene_id': list(missing), 'chromosome_name': ''})\n",
    "ensembl_ids_df = pd.concat([ensembl_ids_df, missing_df])\n",
    "print(len(ensembl_ids_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4e7a37c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to a file\n",
    "ensembl_ids_df.to_csv(path_or_buf = biomart_filename, sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13ef6ec",
   "metadata": {},
   "source": [
    "## Get info on each gene from mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ebd03d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-16000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 16001-17000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 17001-18000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 18001-19000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 19001-20000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 20001-21000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 21001-22000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 22001-23000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 23001-24000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 24001-25000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 25001-26000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 26001-27000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 27001-28000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 28001-29000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 29001-30000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 30001-31000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 31001-32000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 32001-33000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 33001-34000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 34001-35000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 35001-36000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 36001-37000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 37001-38000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 38001-39000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 39001-40000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 40001-41000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 41001-42000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 42001-43000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 43001-44000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 44001-45000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 45001-46000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 46001-47000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 47001-48000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 48001-49000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 49001-50000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 50001-51000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 51001-52000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 52001-53000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 53001-54000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 54001-55000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 55001-56000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 56001-57000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 57001-58000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 58001-59000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 59001-60000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 60001-61000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 61001-62000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 62001-63000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 63001-64000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 64001-65000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 65001-66000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 66001-67000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 67001-68000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 68001-69000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 69001-70000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 70001-71000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 71001-71877...\n",
      "INFO:biothings.client:done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR1</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210082</th>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR2</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000209082</th>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTL1</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  _id  _version      name symbol type_of_gene   alias  \\\n",
       "ensembl_gene_id                                                         \n",
       "ENSG00000210049  4558       2.0  tRNA-Phe   TRNF         tRNA     NaN   \n",
       "ENSG00000211459  4549       2.0    s-rRNA   RNR1         rRNA  MTRNR1   \n",
       "ENSG00000210077  4577       2.0  tRNA-Val   TRNV         tRNA    MTTV   \n",
       "ENSG00000210082  4550       2.0    l-rRNA   RNR2         rRNA  MTRNR2   \n",
       "ENSG00000209082  4567       2.0  tRNA-Leu  TRNL1         tRNA   MTTL1   \n",
       "\n",
       "                                                           summary notfound  \n",
       "ensembl_gene_id                                                              \n",
       "ENSG00000210049                                                NaN      NaN  \n",
       "ENSG00000211459  Enables DNA binding activity and DNA-binding t...      NaN  \n",
       "ENSG00000210077                                                NaN      NaN  \n",
       "ENSG00000210082  Enables G protein-coupled receptor binding act...      NaN  \n",
       "ENSG00000209082  Implicated in cardiomyopathy. [provided by All...      NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "mygene_output = mg.getgenes(ensembl_ids_df['ensembl_gene_id'], \n",
    "                            fields=[\"symbol\", \"name\", \"summary\", \"type_of_gene\", \"alias\"], \n",
    "                            as_dataframe=True)\n",
    "\n",
    "mygene_output.index.rename(\"ensembl_gene_id\", inplace=True)\n",
    "mygene_output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23bb114e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations found for 71469 genes.\n",
      "No annotations found for 1177 genes.\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotations found for \" + str(sum(mygene_output['notfound'].isna())) + \" genes.\")\n",
    "print(\"No annotations found for \" + str(sum(mygene_output['notfound'] == True)) + \" genes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf71470",
   "metadata": {},
   "source": [
    "# Part 2: Clean the data\n",
    "\n",
    "## Join and standardize columns / values\n",
    "\n",
    "For consistency with the `agora-data-tools` transform process, this uses the etl standardize functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "186d8cb8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72648, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR1</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR2</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTL1</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id chromosome_name hgnc_symbol   _id  _version      name  \\\n",
       "0  ENSG00000210049              MT       MT-TF  4558       2.0  tRNA-Phe   \n",
       "1  ENSG00000211459              MT     MT-RNR1  4549       2.0    s-rRNA   \n",
       "2  ENSG00000210077              MT       MT-TV  4577       2.0  tRNA-Val   \n",
       "3  ENSG00000210082              MT     MT-RNR2  4550       2.0    l-rRNA   \n",
       "4  ENSG00000209082              MT      MT-TL1  4567       2.0  tRNA-Leu   \n",
       "\n",
       "  symbol type_of_gene   alias  \\\n",
       "0   TRNF         tRNA     NaN   \n",
       "1   RNR1         rRNA  MTRNR1   \n",
       "2   TRNV         tRNA    MTTV   \n",
       "3   RNR2         rRNA  MTRNR2   \n",
       "4  TRNL1         tRNA   MTTL1   \n",
       "\n",
       "                                             summary notfound  \n",
       "0                                                NaN      NaN  \n",
       "1  Enables DNA binding activity and DNA-binding t...      NaN  \n",
       "2                                                NaN      NaN  \n",
       "3  Enables G protein-coupled receptor binding act...      NaN  \n",
       "4  Implicated in cardiomyopathy. [provided by All...      NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table_merged = pd.merge(left = ensembl_ids_df, right = mygene_output, how = 'left', on = 'ensembl_gene_id')\n",
    "\n",
    "gene_table_merged = utils.standardize_column_names(gene_table_merged)\n",
    "gene_table_merged = utils.standardize_values(gene_table_merged)\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9441115e",
   "metadata": {},
   "source": [
    "## Fix alias field\n",
    "\n",
    "Fix `NaN` values in the `alias` field and make sure every alias value is a list, not a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "285c10d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NaN or NULL alias values become empty lists\n",
    "for row in gene_table_merged.loc[gene_table_merged['alias'].isnull(), 'alias'].index:\n",
    "    gene_table_merged.at[row, 'alias'] = []\n",
    "\n",
    "# Some alias values are a single string, not a list. Turn them into lists here.\n",
    "gene_table_merged['alias'] = gene_table_merged['alias'].apply(lambda cell: cell if isinstance(cell, list) else [cell])\n",
    "\n",
    "# Some alias values are lists of lists or have duplicate values\n",
    "def flatten(row):\n",
    "    flattened = []\n",
    "    for item in row:\n",
    "        if isinstance(item, list):\n",
    "            flattened = flattened + item\n",
    "        else:\n",
    "            flattened.append(item)\n",
    "    return(flattened)\n",
    "\n",
    "gene_table_merged['alias'] = gene_table_merged['alias'].apply(lambda row: list(set(flatten(row))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0be1b92",
   "metadata": {},
   "source": [
    "## Remove duplicate Ensembl IDs from the list. \n",
    "\n",
    "Duplicates in the list typically have the same Ensembl ID but different gene symbols. This usually happens when a single Ensembl ID maps to multiple Entrez IDs in the NCBI database. There's not a good way to reconcile this, so we first use entries whose `hgnc_symbol` from Biomart matches `symbol` from NCBI, then from the remaining duplicates just use the first entry in the list for each ensembl ID and discard the rest, which is what the Agora front end does. The gene symbols of duplicate rows are then added as aliases to the matching unique row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bc63cc53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ENSG00000278294</td>\n",
       "      <td>KI270442.1</td>\n",
       "      <td></td>\n",
       "      <td>124908250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124908250</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENSG00000278294</td>\n",
       "      <td>KI270442.1</td>\n",
       "      <td></td>\n",
       "      <td>124907156</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124907156</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000278294</td>\n",
       "      <td>KI270442.1</td>\n",
       "      <td></td>\n",
       "      <td>124907485</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124907485</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000273730</td>\n",
       "      <td>GL000220.1</td>\n",
       "      <td></td>\n",
       "      <td>124907114</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124907114</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000273730</td>\n",
       "      <td>GL000220.1</td>\n",
       "      <td></td>\n",
       "      <td>124907115</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124907115</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51624</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>285626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51805</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>15</td>\n",
       "      <td>GOLGA6L5P</td>\n",
       "      <td>100133220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>[GOLGA6L3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51806</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>15</td>\n",
       "      <td>GOLGA6L5P</td>\n",
       "      <td>642402</td>\n",
       "      <td>1.0</td>\n",
       "      <td>golgin A6 family like 17, pseudogene</td>\n",
       "      <td>GOLGA6L17P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>[GOLGA6L21P]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71357</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>101928626</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC101928626</td>\n",
       "      <td>LOC101928626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71358</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>124901156</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC124901156</td>\n",
       "      <td>LOC124901156</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>855 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id chromosome_name hgnc_symbol        _id  _version  \\\n",
       "55     ENSG00000278294      KI270442.1              124908250       1.0   \n",
       "56     ENSG00000278294      KI270442.1              124907156       1.0   \n",
       "57     ENSG00000278294      KI270442.1              124907485       1.0   \n",
       "58     ENSG00000273730      GL000220.1              124907114       2.0   \n",
       "59     ENSG00000273730      GL000220.1              124907115       2.0   \n",
       "...                ...             ...         ...        ...       ...   \n",
       "51624  ENSG00000249738               5                 285626       1.0   \n",
       "51805  ENSG00000230373              15   GOLGA6L5P  100133220       1.0   \n",
       "51806  ENSG00000230373              15   GOLGA6L5P     642402       1.0   \n",
       "71357  ENSG00000293331               1              101928626       2.0   \n",
       "71358  ENSG00000293331               1              124901156       2.0   \n",
       "\n",
       "                                       name        symbol type_of_gene  \\\n",
       "55                       5.8S ribosomal RNA  LOC124908250         rRNA   \n",
       "56                       5.8S ribosomal RNA  LOC124907156         rRNA   \n",
       "57                       5.8S ribosomal RNA  LOC124907485         rRNA   \n",
       "58                       5.8S ribosomal RNA  LOC124907114         rRNA   \n",
       "59                       5.8S ribosomal RNA  LOC124907115         rRNA   \n",
       "...                                     ...           ...          ...   \n",
       "51624             uncharacterized LOC285626     LOC285626        ncRNA   \n",
       "51805   golgin A6 family like 3, pseudogene     GOLGA6L3P       pseudo   \n",
       "51806  golgin A6 family like 17, pseudogene    GOLGA6L17P       pseudo   \n",
       "71357          uncharacterized LOC101928626  LOC101928626        ncRNA   \n",
       "71358          uncharacterized LOC124901156  LOC124901156        ncRNA   \n",
       "\n",
       "              alias summary notfound  \n",
       "55               []     NaN      NaN  \n",
       "56               []     NaN      NaN  \n",
       "57               []     NaN      NaN  \n",
       "58               []     NaN      NaN  \n",
       "59               []     NaN      NaN  \n",
       "...             ...     ...      ...  \n",
       "51624            []     NaN      NaN  \n",
       "51805    [GOLGA6L3]     NaN      NaN  \n",
       "51806  [GOLGA6L21P]     NaN      NaN  \n",
       "71357            []     NaN      NaN  \n",
       "71358            []     NaN      NaN  \n",
       "\n",
       "[855 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated() will return true if the ID is a duplicate and is not the first one to appear the list. \n",
    "dupes = gene_table_merged['ensembl_gene_id'].duplicated()\n",
    "dupe_vals = gene_table_merged[dupes]\n",
    "\n",
    "# Rows with duplicated Ensembl IDs\n",
    "all_duplicated = gene_table_merged.loc[gene_table_merged['ensembl_gene_id'].isin(dupe_vals['ensembl_gene_id'])]\n",
    "all_duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "093a2e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83 duplicated genes have been processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71866</th>\n",
       "      <td>ENSG00000283713</td>\n",
       "      <td>HSCHR19KIR_CA01-TB01_CTG3_1</td>\n",
       "      <td>KIR3DP1</td>\n",
       "      <td>128966722</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC128966730, LOC128966731, LOC128966733, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71867</th>\n",
       "      <td>ENSG00000282177</td>\n",
       "      <td>HSCHR15_6_CTG8</td>\n",
       "      <td></td>\n",
       "      <td>124900357</td>\n",
       "      <td>2.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900357</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900358, LOC124900356, LOC124900359]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71868</th>\n",
       "      <td>ENSG00000267635</td>\n",
       "      <td>HSCHR17_5_CTG4</td>\n",
       "      <td></td>\n",
       "      <td>100505724</td>\n",
       "      <td>1.0</td>\n",
       "      <td>keratin associated protein 9-7</td>\n",
       "      <td>KRTAP9-7</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[KAP9.6, KAP9.7, KRTAP9-6, KRTAP9.6, KRTAP9L2,...</td>\n",
       "      <td>Predicted to be located in cytosol. [provided ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71869</th>\n",
       "      <td>ENSG00000288381</td>\n",
       "      <td>HSCHR19KIR_0019-4656-B_CTG3_1</td>\n",
       "      <td></td>\n",
       "      <td>124900572</td>\n",
       "      <td>1.0</td>\n",
       "      <td>killer cell immunoglobulin-like receptor 2DL5B</td>\n",
       "      <td>LOC124900572</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[KIR2DS3, LOC102725023]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71870</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC285626]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71871</th>\n",
       "      <td>ENSG00000288184</td>\n",
       "      <td>HG2525_PATCH</td>\n",
       "      <td></td>\n",
       "      <td>124907358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124907358</td>\n",
       "      <td>LOC124907358</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124907096]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71872</th>\n",
       "      <td>ENSG00000292009</td>\n",
       "      <td>HSCHR1_6_CTG31</td>\n",
       "      <td></td>\n",
       "      <td>124904576</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124904576</td>\n",
       "      <td>LOC124904576</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC128966579, LOC105373277]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71873</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>101928626</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC101928626</td>\n",
       "      <td>LOC101928626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC124901156]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71874</th>\n",
       "      <td>ENSG00000276035</td>\n",
       "      <td>HSCHR19KIR_T7526_A_HAP_CTG3_1</td>\n",
       "      <td>KIR3DP1</td>\n",
       "      <td>128966722</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC128966730, LOC128966731, LOC128966733, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71875</th>\n",
       "      <td>ENSG00000276912</td>\n",
       "      <td>HSCHR19LRC_COX2_CTG3_1</td>\n",
       "      <td>KIR3DP1</td>\n",
       "      <td>128966722</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC128966730, LOC128966731, LOC128966733, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id                chromosome_name hgnc_symbol        _id  \\\n",
       "71866  ENSG00000283713    HSCHR19KIR_CA01-TB01_CTG3_1     KIR3DP1  128966722   \n",
       "71867  ENSG00000282177                 HSCHR15_6_CTG8              124900357   \n",
       "71868  ENSG00000267635                 HSCHR17_5_CTG4              100505724   \n",
       "71869  ENSG00000288381  HSCHR19KIR_0019-4656-B_CTG3_1              124900572   \n",
       "71870  ENSG00000249738                              5              105377683   \n",
       "71871  ENSG00000288184                   HG2525_PATCH              124907358   \n",
       "71872  ENSG00000292009                 HSCHR1_6_CTG31              124904576   \n",
       "71873  ENSG00000293331                              1              101928626   \n",
       "71874  ENSG00000276035  HSCHR19KIR_T7526_A_HAP_CTG3_1     KIR3DP1  128966722   \n",
       "71875  ENSG00000276912         HSCHR19LRC_COX2_CTG3_1     KIR3DP1  128966722   \n",
       "\n",
       "       _version                                               name  \\\n",
       "71866       2.0  putative killer cell immunoglobulin-like recep...   \n",
       "71867       2.0                             U8 small nucleolar RNA   \n",
       "71868       1.0                     keratin associated protein 9-7   \n",
       "71869       1.0     killer cell immunoglobulin-like receptor 2DL5B   \n",
       "71870       1.0                       uncharacterized LOC105377683   \n",
       "71871       1.0                       uncharacterized LOC124907358   \n",
       "71872       1.0                       uncharacterized LOC124904576   \n",
       "71873       2.0                       uncharacterized LOC101928626   \n",
       "71874       2.0  putative killer cell immunoglobulin-like recep...   \n",
       "71875       2.0  putative killer cell immunoglobulin-like recep...   \n",
       "\n",
       "             symbol    type_of_gene  \\\n",
       "71866  LOC128966722  protein-coding   \n",
       "71867  LOC124900357          snoRNA   \n",
       "71868      KRTAP9-7  protein-coding   \n",
       "71869  LOC124900572  protein-coding   \n",
       "71870  LOC105377683           ncRNA   \n",
       "71871  LOC124907358  protein-coding   \n",
       "71872  LOC124904576           ncRNA   \n",
       "71873  LOC101928626           ncRNA   \n",
       "71874  LOC128966722  protein-coding   \n",
       "71875  LOC128966722  protein-coding   \n",
       "\n",
       "                                                   alias  \\\n",
       "71866  [LOC128966730, LOC128966731, LOC128966733, LOC...   \n",
       "71867         [LOC124900358, LOC124900356, LOC124900359]   \n",
       "71868  [KAP9.6, KAP9.7, KRTAP9-6, KRTAP9.6, KRTAP9L2,...   \n",
       "71869                            [KIR2DS3, LOC102725023]   \n",
       "71870                                        [LOC285626]   \n",
       "71871                                     [LOC124907096]   \n",
       "71872                       [LOC128966579, LOC105373277]   \n",
       "71873                                     [LOC124901156]   \n",
       "71874  [LOC128966730, LOC128966731, LOC128966733, LOC...   \n",
       "71875  [LOC128966730, LOC128966731, LOC128966733, LOC...   \n",
       "\n",
       "                                                 summary notfound  \n",
       "71866                                                NaN      NaN  \n",
       "71867                                                NaN      NaN  \n",
       "71868  Predicted to be located in cytosol. [provided ...      NaN  \n",
       "71869                                                NaN      NaN  \n",
       "71870                                                NaN      NaN  \n",
       "71871                                                NaN      NaN  \n",
       "71872                                                NaN      NaN  \n",
       "71873                                                NaN      NaN  \n",
       "71874                                                NaN      NaN  \n",
       "71875                                                NaN      NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_dupes = set(gene_table_merged.index) - set(all_duplicated.index)\n",
    "keep_df = gene_table_merged.loc[non_dupes].copy(deep=True)\n",
    "\n",
    "# For each duplicated Ensembl ID, collapse to 1 row and append that row to keep_df\n",
    "for ens_id in set(all_duplicated['ensembl_gene_id']):\n",
    "    group = all_duplicated.loc[all_duplicated['ensembl_gene_id'] == ens_id].copy(deep=True)\n",
    "    matches = group['hgnc_symbol'] == group['symbol']\n",
    "    \n",
    "    # If there is a single entry with a matching symbol from NCBI, use that row and ignore the others\n",
    "    if (sum(matches) == 1):\n",
    "        keep_df = keep_df.append(group.loc[matches], ignore_index=True)\n",
    "    # Multiple or no matching symbols, save the first entry in the list and add the other rows as aliases\n",
    "    else:\n",
    "        # For multiple matching symbols, discard non-matching entries and continue\n",
    "        if sum(matches) > 1: \n",
    "            group = group.loc[matches]\n",
    "        \n",
    "        # Add all duplicate symbols and their aliases to the alias field of the first entry\n",
    "        for row in group.index[1:]:\n",
    "            group.at[group.index[0], 'alias'].append(group['symbol'][row])\n",
    "            if (len(group.at[row, 'alias']) > 0):\n",
    "                group.at[group.index[0], 'alias'] = group.at[group.index[0], 'alias'] + group['alias'][row]\n",
    "        \n",
    "        # Make sure we didn't add duplicate aliases\n",
    "        group.at[group.index[0], 'alias'] = list(set(group.at[group.index[0], 'alias']))\n",
    "\n",
    "        # Keep the first row only, which now has all the aliases\n",
    "        keep_df = keep_df.append(group.iloc[0], ignore_index=True)\n",
    "        \n",
    "print(str(len(all_duplicated.drop_duplicates('ensembl_gene_id'))) + \" duplicated genes have been processed.\")\n",
    "gene_table_merged = keep_df.reset_index(drop=True)\n",
    "gene_table_merged.tail(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "702c361d",
   "metadata": {},
   "source": [
    "# Part 3: Create Ensembl archive permalinks\n",
    "\n",
    "## Get a table of Ensembl archive URLs\n",
    "\n",
    "This is where we need to use the R biomaRt library specifically, instead of any of the available Python interfaces to Biomart, to get a table of Ensembl release versions and their corresponding archive URLs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a1bbdee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             name     date                                 url version\n",
      "1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n",
      "2     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111\n",
      "3     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110\n",
      "4     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n",
      "5     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n",
      "6     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n",
      "7     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n",
      "8     Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n",
      "9     Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n",
      "10    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n",
      "11    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n",
      "12    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n",
      "13    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n",
      "14     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n",
      "15     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n",
      "16     Ensembl 97 Jul 2019 https://jul2019.archive.ensembl.org      97\n",
      "17     Ensembl 96 Apr 2019 https://apr2019.archive.ensembl.org      96\n",
      "18     Ensembl 95 Jan 2019 https://jan2019.archive.ensembl.org      95\n",
      "19     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80\n",
      "20     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n",
      "21     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n",
      "22     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n",
      "   current_release\n",
      "1                 \n",
      "2                *\n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10                \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n",
      "17                \n",
      "18                \n",
      "19                \n",
      "20                \n",
      "21                \n",
      "22                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "archive_df = r.listEnsemblArchives()\n",
    "archive_df.to_csvfile(path = archive_filename, row_names = False, quote = False)\n",
    "\n",
    "print(archive_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e150d8e",
   "metadata": {},
   "source": [
    "## Query Ensembl for each gene's version\n",
    "\n",
    "Ensembl's REST API can only take 1000 genes at once, so this is looped to query groups of 1000. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9a747309",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71876\n",
      "Querying genes 1 - 1000\n",
      "Querying genes 1001 - 2000\n",
      "Querying genes 2001 - 3000\n",
      "Querying genes 3001 - 4000\n",
      "Querying genes 4001 - 5000\n",
      "Querying genes 5001 - 6000\n",
      "Querying genes 6001 - 7000\n",
      "Querying genes 7001 - 8000\n",
      "Querying genes 8001 - 9000\n",
      "Querying genes 9001 - 10000\n",
      "Querying genes 10001 - 11000\n",
      "Querying genes 11001 - 12000\n",
      "Querying genes 12001 - 13000\n",
      "Querying genes 13001 - 14000\n",
      "Querying genes 14001 - 15000\n",
      "Querying genes 15001 - 16000\n",
      "Querying genes 16001 - 17000\n",
      "Querying genes 17001 - 18000\n",
      "Querying genes 18001 - 19000\n",
      "Querying genes 19001 - 20000\n",
      "Querying genes 20001 - 21000\n",
      "Querying genes 21001 - 22000\n",
      "Querying genes 22001 - 23000\n",
      "Querying genes 23001 - 24000\n",
      "Querying genes 24001 - 25000\n",
      "Querying genes 25001 - 26000\n",
      "Querying genes 26001 - 27000\n",
      "Querying genes 27001 - 28000\n",
      "Querying genes 28001 - 29000\n",
      "Querying genes 29001 - 30000\n",
      "Querying genes 30001 - 31000\n",
      "Querying genes 31001 - 32000\n",
      "Querying genes 32001 - 33000\n",
      "Querying genes 33001 - 34000\n",
      "Querying genes 34001 - 35000\n",
      "Querying genes 35001 - 36000\n",
      "Querying genes 36001 - 37000\n",
      "Querying genes 37001 - 38000\n",
      "Querying genes 38001 - 39000\n",
      "Querying genes 39001 - 40000\n",
      "Querying genes 40001 - 41000\n",
      "Querying genes 41001 - 42000\n",
      "Querying genes 42001 - 43000\n",
      "Querying genes 43001 - 44000\n",
      "Querying genes 44001 - 45000\n",
      "Querying genes 45001 - 46000\n",
      "Querying genes 46001 - 47000\n",
      "Querying genes 47001 - 48000\n",
      "Querying genes 48001 - 49000\n",
      "Querying genes 49001 - 50000\n",
      "Querying genes 50001 - 51000\n",
      "Querying genes 51001 - 52000\n",
      "Querying genes 52001 - 53000\n",
      "Querying genes 53001 - 54000\n",
      "Querying genes 54001 - 55000\n",
      "Querying genes 55001 - 56000\n",
      "Querying genes 56001 - 57000\n",
      "Querying genes 57001 - 58000\n",
      "Querying genes 58001 - 59000\n",
      "Querying genes 59001 - 60000\n",
      "Querying genes 60001 - 61000\n",
      "Querying genes 61001 - 62000\n",
      "Querying genes 62001 - 63000\n",
      "Querying genes 63001 - 64000\n",
      "Querying genes 64001 - 65000\n",
      "Querying genes 65001 - 66000\n",
      "Querying genes 66001 - 67000\n",
      "Querying genes 67001 - 68000\n",
      "Querying genes 68001 - 69000\n",
      "Querying genes 69001 - 70000\n",
      "Querying genes 70001 - 71000\n",
      "Querying genes 71001 - 71876\n",
      "71876\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>release</th>\n",
       "      <th>peptide</th>\n",
       "      <th>version</th>\n",
       "      <th>type</th>\n",
       "      <th>assembly</th>\n",
       "      <th>latest</th>\n",
       "      <th>id</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>is_current</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71871</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000288184.1</td>\n",
       "      <td>ENSG00000288184</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71872</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000292009.1</td>\n",
       "      <td>ENSG00000292009</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71873</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000293331.1</td>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71874</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000276035.1</td>\n",
       "      <td>ENSG00000276035</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71875</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000276912.1</td>\n",
       "      <td>ENSG00000276912</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      release peptide  version  type assembly             latest  \\\n",
       "71871     111    None        1  Gene   GRCh38  ENSG00000288184.1   \n",
       "71872     111    None        1  Gene   GRCh38  ENSG00000292009.1   \n",
       "71873     111    None        1  Gene   GRCh38  ENSG00000293331.1   \n",
       "71874     111    None        1  Gene   GRCh38  ENSG00000276035.1   \n",
       "71875     111    None        1  Gene   GRCh38  ENSG00000276912.1   \n",
       "\n",
       "                    id possible_replacement is_current  \n",
       "71871  ENSG00000288184                   []          1  \n",
       "71872  ENSG00000292009                   []          1  \n",
       "71873  ENSG00000293331                   []          1  \n",
       "71874  ENSG00000276035                   []          1  \n",
       "71875  ENSG00000276912                   []          1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://rest.ensembl.org/archive/id\"\n",
    "headers = {\"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "\n",
    "ids = gene_table_merged['ensembl_gene_id'].tolist()\n",
    "print(len(ids))\n",
    "\n",
    "# We can only query 1000 genes at a time\n",
    "batch_ind = range(0, len(ids), 1000)\n",
    "results = []\n",
    "\n",
    "for B in batch_ind:\n",
    "    end = min(len(ids), B + 1000)\n",
    "    print(\"Querying genes \" + str(B+1) + \" - \" + str(end))\n",
    "    \n",
    "    request_data = '{ \"id\" : ' + str(ids[B:end]) + ' }'\n",
    "    request_data = request_data.replace(\"'\", \"\\\"\")\n",
    "    \n",
    "    ok = False\n",
    "    tries = 0\n",
    "    \n",
    "    while tries < 5 and not ok:\n",
    "        try:\n",
    "            res = requests.post(url, headers=headers, data=request_data)\n",
    "            ok = res.ok\n",
    "        except:\n",
    "            ok = False\n",
    "        \n",
    "        tries = tries + 1\n",
    "        \n",
    "        if not ok:\n",
    "            #res.raise_for_status()\n",
    "            print(\"Error retrieving Ensembl versions for genes \" + str(B+1) + \" - \" + str(end) + \n",
    "                  \". Trying again...\")\n",
    "        else:\n",
    "            results = results + res.json()\n",
    "            break\n",
    "\n",
    "print(len(results))\n",
    "\n",
    "versions = pd.json_normalize(results)\n",
    "\n",
    "versions.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5c108238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release\n",
       "100       22\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       34\n",
       "107       10\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111    70711\n",
       "80        21\n",
       "81         2\n",
       "82        10\n",
       "84       673\n",
       "87        61\n",
       "89        20\n",
       "91        75\n",
       "93        53\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         9\n",
       "99         7\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions.groupby('release').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bf5aecb1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71876\n",
      "71876\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Check that all IDs are the same between the result and the gene table\n",
    "print(len(versions['id']))\n",
    "print(len(gene_table_merged))\n",
    "print(all(versions['id'].isin(gene_table_merged['ensembl_gene_id'])) and \n",
    "      all(gene_table_merged['ensembl_gene_id'].isin(versions['id'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7fc8bbcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure everything is GRCh38, not GRCh37\n",
    "all(versions['assembly'] == \"GRCh38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e29252",
   "metadata": {},
   "source": [
    "## Create permalinks based on archive version\n",
    "\n",
    "**Not all of these versions have an archive.** We can go back to the closest previous archive for these but the link isn't guaranteed to work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0d5b5652",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "archive_table = pd.read_csv(archive_filename)\n",
    "\n",
    "# Remove GRCh37 from the archive list\n",
    "archive_table = archive_table[archive_table['version'] != \"GRCh37\"].reset_index()\n",
    "\n",
    "archive_table['numeric_version'] = archive_table['version'].astype(int)\n",
    "\n",
    "def closest_release(release, archive_table):\n",
    "    if release in archive_table:\n",
    "        return release\n",
    "    \n",
    "    return max([V for V in archive_table['numeric_version'] if V <= release])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "337b2890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "closest_release\n",
       "80       915\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         9\n",
       "99         7\n",
       "100       22\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       34\n",
       "107       10\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111    70711\n",
       "dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions['closest_release'] = 0\n",
    "\n",
    "releases = versions['release'].drop_duplicates().astype(int)\n",
    "\n",
    "# Only have to call closest_release once per version, instead of >70k times\n",
    "for release in releases:\n",
    "    versions.loc[versions['release'] == str(release), 'closest_release'] = closest_release(release, archive_table)\n",
    "    \n",
    "versions.groupby('closest_release').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "343e5006",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>release</th>\n",
       "      <th>peptide</th>\n",
       "      <th>version</th>\n",
       "      <th>type</th>\n",
       "      <th>assembly</th>\n",
       "      <th>latest</th>\n",
       "      <th>id</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>111</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  release peptide  version  type assembly             latest               id  \\\n",
       "0     111    None        1  Gene   GRCh38  ENSG00000210049.1  ENSG00000210049   \n",
       "1     111    None        2  Gene   GRCh38  ENSG00000211459.2  ENSG00000211459   \n",
       "2     111    None        1  Gene   GRCh38  ENSG00000210077.1  ENSG00000210077   \n",
       "3     111    None        2  Gene   GRCh38  ENSG00000210082.2  ENSG00000210082   \n",
       "4     111    None        1  Gene   GRCh38  ENSG00000209082.1  ENSG00000209082   \n",
       "\n",
       "  possible_replacement is_current  closest_release  \\\n",
       "0                   []          1              111   \n",
       "1                   []          1              111   \n",
       "2                   []          1              111   \n",
       "3                   []          1              111   \n",
       "4                   []          1              111   \n",
       "\n",
       "                                           permalink  \n",
       "0  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4  https://jan2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions['permalink'] = ''\n",
    "\n",
    "for i in versions.index:\n",
    "    match = archive_table['numeric_version'] == versions.at[i, 'closest_release']\n",
    "    url = archive_table.loc[match, 'url'].to_string(index = False)\n",
    "    if len(url) > 0:\n",
    "        versions.at[i, 'permalink'] = url + \"/Homo_sapiens/Gene/Summary?db=core;g=\" + versions.at[i, 'id']\n",
    "\n",
    "versions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4b01719d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>release</th>\n",
       "      <th>peptide</th>\n",
       "      <th>version</th>\n",
       "      <th>type</th>\n",
       "      <th>assembly</th>\n",
       "      <th>latest</th>\n",
       "      <th>id</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70629</th>\n",
       "      <td>84</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000265650.1</td>\n",
       "      <td>ENSG00000265650</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70630</th>\n",
       "      <td>84</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000265009.1</td>\n",
       "      <td>ENSG00000265009</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70632</th>\n",
       "      <td>84</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000221720.2</td>\n",
       "      <td>ENSG00000221720</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70633</th>\n",
       "      <td>84</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000274325.1</td>\n",
       "      <td>ENSG00000274325</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70634</th>\n",
       "      <td>84</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000278612.1</td>\n",
       "      <td>ENSG00000278612</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      release peptide  version  type assembly             latest  \\\n",
       "70629      84    None        1  Gene   GRCh38  ENSG00000265650.1   \n",
       "70630      84    None        1  Gene   GRCh38  ENSG00000265009.1   \n",
       "70632      84    None        2  Gene   GRCh38  ENSG00000221720.2   \n",
       "70633      84    None        1  Gene   GRCh38  ENSG00000274325.1   \n",
       "70634      84    None        1  Gene   GRCh38  ENSG00000278612.1   \n",
       "\n",
       "                    id possible_replacement is_current  closest_release  \\\n",
       "70629  ENSG00000265650                   []                          80   \n",
       "70630  ENSG00000265009                   []                          80   \n",
       "70632  ENSG00000221720                   []                          80   \n",
       "70633  ENSG00000274325                   []                          80   \n",
       "70634  ENSG00000278612                   []                          80   \n",
       "\n",
       "                                               permalink  \n",
       "70629  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "70630  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "70632  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "70633  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "70634  https://may2015.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[versions['closest_release'] < 100].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c4128cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://jan2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000210049\n",
      "https://jan2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000210174\n"
     ]
    }
   ],
   "source": [
    "print(versions['permalink'][0])\n",
    "print(versions['permalink'][25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "73791e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Does every gene have an associated URL?\n",
    "url_base_len = len(archive_table['url'][0]) + 1\n",
    "all([len(url) > url_base_len for url in versions['permalink']])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b98b62a",
   "metadata": {},
   "source": [
    "# Part 4: Add permalinks to the gene table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f3edfd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(71876, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[MTRNR1]</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[MTTV]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[MTRNR2]</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[MTTL1]</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id chromosome_name hgnc_symbol   _id  _version      name  \\\n",
       "0  ENSG00000210049              MT       MT-TF  4558       2.0  tRNA-Phe   \n",
       "1  ENSG00000211459              MT     MT-RNR1  4549       2.0    s-rRNA   \n",
       "2  ENSG00000210077              MT       MT-TV  4577       2.0  tRNA-Val   \n",
       "3  ENSG00000210082              MT     MT-RNR2  4550       2.0    l-rRNA   \n",
       "4  ENSG00000209082              MT      MT-TL1  4567       2.0  tRNA-Leu   \n",
       "\n",
       "  symbol type_of_gene     alias  \\\n",
       "0   TRNF         tRNA        []   \n",
       "1   RNR1         rRNA  [MTRNR1]   \n",
       "2   TRNV         tRNA    [MTTV]   \n",
       "3   RNR2         rRNA  [MTRNR2]   \n",
       "4  TRNL1         tRNA   [MTTL1]   \n",
       "\n",
       "                                             summary notfound ensembl_release  \\\n",
       "0                                                NaN      NaN             111   \n",
       "1  Enables DNA binding activity and DNA-binding t...      NaN             111   \n",
       "2                                                NaN      NaN             111   \n",
       "3  Enables G protein-coupled receptor binding act...      NaN             111   \n",
       "4  Implicated in cardiomyopathy. [provided by All...      NaN             111   \n",
       "\n",
       "  possible_replacement                                          permalink  \n",
       "0                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions = versions[['id', 'release', 'possible_replacement', 'permalink']]\n",
    "versions.rename(columns={'id': 'ensembl_gene_id', 'release': 'ensembl_release'}, inplace=True)\n",
    "\n",
    "gene_table_merged = pd.merge(left = gene_table_merged, right = versions, how = 'left', on = 'ensembl_gene_id')\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01eba7c2",
   "metadata": {},
   "source": [
    "### Final cleanup\n",
    "Unfilled \"possible_replacement\" entries should be changed from NaN to empty lists. \n",
    "\n",
    "\"possible_replacement\" entries that have data in them exist as a list of dicts, and need to have the Ensembl IDs pulled out of them as a list of strings. \n",
    "\n",
    "Remove unneeded columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d0c07b7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>[MTRNR1]</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>[MTTV]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>[MTRNR2]</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>[MTTL1]</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71871</th>\n",
       "      <td>ENSG00000288184</td>\n",
       "      <td>uncharacterized LOC124907358</td>\n",
       "      <td>[LOC124907096]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC124907358</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71872</th>\n",
       "      <td>ENSG00000292009</td>\n",
       "      <td>uncharacterized LOC124904576</td>\n",
       "      <td>[LOC128966579, LOC105373277]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC124904576</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71873</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>uncharacterized LOC101928626</td>\n",
       "      <td>[LOC124901156]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC101928626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71874</th>\n",
       "      <td>ENSG00000276035</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>[LOC128966730, LOC128966731, LOC128966733, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71875</th>\n",
       "      <td>ENSG00000276912</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>[LOC128966730, LOC128966731, LOC128966733, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71876 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id                                               name  \\\n",
       "0      ENSG00000210049                                           tRNA-Phe   \n",
       "1      ENSG00000211459                                             s-rRNA   \n",
       "2      ENSG00000210077                                           tRNA-Val   \n",
       "3      ENSG00000210082                                             l-rRNA   \n",
       "4      ENSG00000209082                                           tRNA-Leu   \n",
       "...                ...                                                ...   \n",
       "71871  ENSG00000288184                       uncharacterized LOC124907358   \n",
       "71872  ENSG00000292009                       uncharacterized LOC124904576   \n",
       "71873  ENSG00000293331                       uncharacterized LOC101928626   \n",
       "71874  ENSG00000276035  putative killer cell immunoglobulin-like recep...   \n",
       "71875  ENSG00000276912  putative killer cell immunoglobulin-like recep...   \n",
       "\n",
       "                                                   alias  \\\n",
       "0                                                     []   \n",
       "1                                               [MTRNR1]   \n",
       "2                                                 [MTTV]   \n",
       "3                                               [MTRNR2]   \n",
       "4                                                [MTTL1]   \n",
       "...                                                  ...   \n",
       "71871                                     [LOC124907096]   \n",
       "71872                       [LOC128966579, LOC105373277]   \n",
       "71873                                     [LOC124901156]   \n",
       "71874  [LOC128966730, LOC128966731, LOC128966733, LOC...   \n",
       "71875  [LOC128966730, LOC128966731, LOC128966733, LOC...   \n",
       "\n",
       "                                                 summary        symbol  \\\n",
       "0                                                    NaN          TRNF   \n",
       "1      Enables DNA binding activity and DNA-binding t...          RNR1   \n",
       "2                                                    NaN          TRNV   \n",
       "3      Enables G protein-coupled receptor binding act...          RNR2   \n",
       "4      Implicated in cardiomyopathy. [provided by All...         TRNL1   \n",
       "...                                                  ...           ...   \n",
       "71871                                                NaN  LOC124907358   \n",
       "71872                                                NaN  LOC124904576   \n",
       "71873                                                NaN  LOC101928626   \n",
       "71874                                                NaN  LOC128966722   \n",
       "71875                                                NaN  LOC128966722   \n",
       "\n",
       "         type_of_gene ensembl_release possible_replacement  \\\n",
       "0                tRNA             111                   []   \n",
       "1                rRNA             111                   []   \n",
       "2                tRNA             111                   []   \n",
       "3                rRNA             111                   []   \n",
       "4                tRNA             111                   []   \n",
       "...               ...             ...                  ...   \n",
       "71871  protein-coding             111                   []   \n",
       "71872           ncRNA             111                   []   \n",
       "71873           ncRNA             111                   []   \n",
       "71874  protein-coding             111                   []   \n",
       "71875  protein-coding             111                   []   \n",
       "\n",
       "                                               permalink  \n",
       "0      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "...                                                  ...  \n",
       "71871  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "71872  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "71873  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "71874  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "71875  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "\n",
       "[71876 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for row in gene_table_merged.loc[gene_table_merged['possible_replacement'].isnull(), 'possible_replacement'].index:\n",
    "    gene_table_merged.at[row, 'possible_replacement'] = []\n",
    "    \n",
    "gene_table_merged['possible_replacement'] = gene_table_merged.apply(\n",
    "    lambda row: row['possible_replacement'] if len(row['possible_replacement']) == 0\n",
    "        else [x['stable_id'] for x in row['possible_replacement']],\n",
    "        axis=1\n",
    ")\n",
    "\n",
    "gene_table_merged = gene_table_merged[['ensembl_gene_id', 'name', 'alias', 'summary', 'symbol', \n",
    "                                       'type_of_gene', 'ensembl_release', 'possible_replacement', 'permalink']]\n",
    "\n",
    "gene_table_merged\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b371435",
   "metadata": {},
   "source": [
    "### Write to a file\n",
    "This will get uploaded to Synapse as [syn25953363](https://www.synapse.org/#!Synapse:syn25953363)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f2287922",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_merged = gene_table_merged[['ensembl_gene_id', 'name', 'symbol', 'type_of_gene', 'alias',\n",
    "                                       'summary', 'ensembl_release', 'possible_replacement', 'permalink']]\n",
    "gene_table_merged = gene_table_merged.sort_values(by = 'ensembl_gene_id').reset_index(drop=True)\n",
    "gene_table_merged\n",
    "gene_table_merged.to_feather('../../output/gene_table_merged_GRCh38.p14.feather')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
