{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d9f808",
   "metadata": {},
   "source": [
    "# Preprocess Gene Annotations\n",
    "\n",
    "This notebook creates a table of gene annotations by:\n",
    "1. Querying Biomart for all Ensembl IDs in the database\n",
    "2. Querying MyGene for annotation about those IDs\n",
    "3. Querying Ensembl for the most recent Ensembl release for each ID\n",
    "4. Building a permalink to the Ensembl archive page for each ID\n",
    "\n",
    "This gene annotation table is read in by `agoradataprocessing/process.py` to be used in the `gene_info` transformation. \n",
    "\n",
    "***Note:*** *This notebook is exploratory and should eventually be converted to a Python script that is run through an automated process.*\n",
    "\n",
    "## Installation requirements\n",
    "\n",
    "#### Linux / Windows / Mac\n",
    "\n",
    "Install R: https://cran.r-project.org/\n",
    "\n",
    "Install Python and agora-data-tools following the instructions in this repository's README. This notebook assumes it is being run from the same `pipenv` virtual environment as agora-data-tools. \n",
    "\n",
    "Then install the following packages using `pip`:\n",
    "```\n",
    "pip install rpy2 mygene\n",
    "```\n",
    "\n",
    "#### Note for Macs with M1 chips (2020 and newer)\n",
    "\n",
    "Install as above, but make sure that your R installation is the arm64 version (R-4.X.X-arm64.pkg) so that the architecture matches what pip is using. \n",
    "You may also need to install an older version of `rpy2` on the Mac:\n",
    "```\n",
    "pip install rpy2==3.5.12\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2dc92c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocManager\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.0 (2023-04-21)\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version '3.17' is out-of-date; the current release version '3.18'\n",
      "  is available with R version '4.3'; see https://bioconductor.org/install\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: biomaRt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from rpy2.robjects import r\n",
    "import pandas as pd\n",
    "import mygene\n",
    "import numpy as np\n",
    "import requests\n",
    "import agoradatatools.etl.utils as utils\n",
    "import agoradatatools.etl.extract as extract\n",
    "import preprocessing_utils\n",
    "\n",
    "r('if (!require(\"BiocManager\", character.only = TRUE)) { install.packages(\"BiocManager\") }')\n",
    "r('if (!require(\"biomaRt\")) { BiocManager::install(\"biomaRt\") }')\n",
    "\n",
    "r.library(\"biomaRt\")\n",
    "\n",
    "biomart_filename = \"../../output/biomart_ensg_list.txt\"\n",
    "archive_filename = \"../../output/ensembl_archive_list.csv\"\n",
    "config_filename = \"../../../../config.yaml\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec36f54c",
   "metadata": {},
   "source": [
    "# Part 1: Get gene annotation data\n",
    "\n",
    "## Query Biomart for a list of all Ensembl IDs in the database of human genes. \n",
    "\n",
    "Here we use the R library `biomaRt`. There is no canonical Python library with the features we need for this script. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d1e115d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       ensembl_gene_id chromosome_name hgnc_symbol\n",
      "0      ENSG00000210049              MT       MT-TF\n",
      "1      ENSG00000211459              MT     MT-RNR1\n",
      "2      ENSG00000210077              MT       MT-TV\n",
      "3      ENSG00000210082              MT     MT-RNR2\n",
      "4      ENSG00000209082              MT      MT-TL1\n",
      "...                ...             ...         ...\n",
      "70707  ENSG00000288629               1            \n",
      "70708  ENSG00000288678               1            \n",
      "70709  ENSG00000290825               1            \n",
      "70710  ENSG00000227232               1      WASH7P\n",
      "70711  ENSG00000290826               1            \n",
      "\n",
      "[70712 rows x 3 columns]\n",
      "70712 genes found.\n"
     ]
    }
   ],
   "source": [
    "# Sometimes Biomart doesn't respond and the command needs to be sent again. Try up to 5 times.\n",
    "for T in range(5):\n",
    "    try:\n",
    "        mart = r.useEnsembl(biomart=\"ensembl\", dataset=\"hsapiens_gene_ensembl\")\n",
    "        ensembl_ids = r.getBM(\n",
    "            attributes=r.c(\"ensembl_gene_id\", \"chromosome_name\", \"hgnc_symbol\"),\n",
    "            mart=mart,\n",
    "            useCache=False,\n",
    "        )\n",
    "\n",
    "    except:\n",
    "        print(\"Trying again...\")\n",
    "        ensembl_ids = None\n",
    "    else:\n",
    "        break\n",
    "\n",
    "if ensembl_ids is None or ensembl_ids.nrow == 0:\n",
    "    print(\"Biomart was unresponsive after 5 attempts. Try again later.\")\n",
    "\n",
    "else:\n",
    "    # Convert the ensembl_gene_id column from R object to a python list\n",
    "    ensembl_ids_df = pd.DataFrame(\n",
    "        {\n",
    "            \"ensembl_gene_id\": list(ensembl_ids.rx2(\"ensembl_gene_id\")),\n",
    "            \"chromosome_name\": list(ensembl_ids.rx2(\"chromosome_name\")),\n",
    "            \"hgnc_symbol\": list(ensembl_ids.rx2(\"hgnc_symbol\")),\n",
    "        }\n",
    "    )\n",
    "    print(ensembl_ids_df)\n",
    "    print(str(ensembl_ids_df.shape[0]) + \" genes found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ddd4e52",
   "metadata": {},
   "source": [
    "## Remove HASGs from Biomart results\n",
    "This removes all human alternative sequence genes (HASGs) and patches from the Biomart results. These genes can be identified by their chromosome name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1709e32c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63188 genes remaining after HASG filtering.\n"
     ]
    }
   ],
   "source": [
    "ensembl_ids_df = preprocessing_utils.filter_hasgs(\n",
    "    df=ensembl_ids_df, chromosome_name_column=\"chromosome_name\"\n",
    ")\n",
    "print(str(ensembl_ids_df.shape[0]) + \" genes remaining after HASG filtering.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f494e58f",
   "metadata": {},
   "source": [
    "## Add Ensembl IDs from data sets that will be processed by agora-data-tools\n",
    "\n",
    "Some of these datasets have older/retired Ensembl IDs that no longer exist in the current Ensembl database. Loop through all data sets in the config file to get all Ensembl IDs they use, and add missing ones to the gene list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a3fdbeec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes_biodomains': ('syn44151254.4', 'csv'),\n",
       " 'neuropath_regression_results': ('syn22017882.5', 'csv'),\n",
       " 'proteomics': ('syn18689335.3', 'csv'),\n",
       " 'proteomics_tmt': ('syn35221005.2', 'csv'),\n",
       " 'proteomics_srm': ('syn52579640.4', 'csv'),\n",
       " 'target_exp_validation_harmonized': ('syn24184512.8', 'csv'),\n",
       " 'metabolomics': ('syn26064497.1', 'feather'),\n",
       " 'igap': ('syn12514826.5', 'csv'),\n",
       " 'eqtl': ('syn12514912.3', 'csv'),\n",
       " 'diff_exp_data': ('syn27211942.1', 'tsv'),\n",
       " 'target_list': ('syn12540368.47', 'csv'),\n",
       " 'median_expression': ('syn27211878.2', 'csv'),\n",
       " 'druggability': ('syn13363443.11', 'csv'),\n",
       " 'tep_adi_info': ('syn51942280.2', 'csv'),\n",
       " 'team_info': ('syn12615624.18', 'csv'),\n",
       " 'team_member_info': ('syn12615633.18', 'csv'),\n",
       " 'overall_scores': ('syn25575156.13', 'table'),\n",
       " 'networks': ('syn11685347.1', 'csv')}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = utils._get_config(config_path=config_filename)\n",
    "datasets = config[\"datasets\"]\n",
    "\n",
    "files = {}\n",
    "\n",
    "for dataset in datasets:\n",
    "    dataset_name = list(dataset.keys())[0]\n",
    "\n",
    "    for entity in dataset[dataset_name][\"files\"]:\n",
    "        entity_id = entity[\"id\"]\n",
    "        entity_format = entity[\"format\"]\n",
    "        entity_name = entity[\"name\"]\n",
    "\n",
    "        # Ignore json files, which are post-processed and not what we're interested in.\n",
    "        # Also ignore \"gene_metadata\" since that's the file we're making here.\n",
    "        if entity_format != \"json\" and entity_name != \"gene_metadata\":\n",
    "            files[entity_name] = (entity_id, entity_format)\n",
    "\n",
    "# There are some duplicate synID's in this list but that doesn't really matter\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1a2120",
   "metadata": {},
   "source": [
    "### We should now have a list of all raw data files ingested. Get each one and add its Ensembl IDs to the gene list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9843689d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:synapseclient_default:Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target_exp_validation_harmonized has an n/A Ensembl ID\n",
      "WARNING: no Ensembl ID column found for team_info!\n",
      "WARNING: no Ensembl ID column found for team_member_info!\n"
     ]
    }
   ],
   "source": [
    "# Assumes you have already logged in with a valid token\n",
    "syn = utils._login_to_synapse(token=None)\n",
    "\n",
    "# The various column names used to store Ensembl IDs in the files\n",
    "col_names = [\"ENSG\", \"ensembl_gene_id\", \"GeneID\", \"ensembl_id\"]\n",
    "file_ensembl_list = []\n",
    "\n",
    "for file in files.keys():\n",
    "    df = extract.get_entity_as_df(syn_id=files[file][0], source=files[file][1], syn=syn)\n",
    "\n",
    "    file_ensembl_ids = None\n",
    "\n",
    "    for C in col_names:\n",
    "        if C in df.columns:\n",
    "            file_ensembl_ids = df[C]\n",
    "\n",
    "    # networks file is a special case\n",
    "    if file == \"networks\":\n",
    "        file_ensembl_ids = pd.melt(\n",
    "            df[[\"geneA_ensembl_gene_id\", \"geneB_ensembl_gene_id\"]]\n",
    "        )[\"value\"]\n",
    "\n",
    "    # genes_biodomains is a special case -- the ensembl_id field has some semicolon-separated lists in it\n",
    "    if file == \"genes_biodomains\":\n",
    "        df = df[[\"Biodomain\", \"ensembl_id\"]].drop_duplicates().dropna()\n",
    "        df = utils.split_delimited_field_to_multiple_rows(\n",
    "            df=df, split_field=\"ensembl_id\", delim=\";\"\n",
    "        )\n",
    "        file_ensembl_ids = df[\"ensembl_id\"].drop_duplicates()\n",
    "\n",
    "    if file_ensembl_ids is not None:\n",
    "        file_ensembl_list = file_ensembl_list + file_ensembl_ids.tolist()\n",
    "        if \"n/A\" in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an n/A Ensembl ID\")\n",
    "        if np.NaN in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an NaN Ensembl ID\")\n",
    "    else:\n",
    "        print(\"WARNING: no Ensembl ID column found for \" + file + \"!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1303e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1821 genes from the data files are missing from Biomart results and will be added.\n",
      "65009\n"
     ]
    }
   ],
   "source": [
    "file_ensembl_list = list(set(file_ensembl_list))\n",
    "file_ensembl_list.remove(\"n/A\")  # Necessary because one data set has an Ensembl ID set to \"n/A\"\n",
    "\n",
    "# Add Ensembl IDs that are in the files but not in the biomart result\n",
    "missing = set(file_ensembl_list) - set(ensembl_ids_df[\"ensembl_gene_id\"])\n",
    "print(\n",
    "    str(len(missing))\n",
    "    + \" genes from the data files are missing from Biomart results and will be added.\"\n",
    ")\n",
    "\n",
    "missing_df = pd.DataFrame({\"ensembl_gene_id\": list(missing), \"chromosome_name\": \"\"})\n",
    "ensembl_ids_df = pd.concat([ensembl_ids_df, missing_df])\n",
    "ensembl_ids_df = ensembl_ids_df.dropna(subset=[\"ensembl_gene_id\"])\n",
    "print(len(ensembl_ids_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4e7a37c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to a file\n",
    "ensembl_ids_df.to_csv(path_or_buf=biomart_filename, sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13ef6ec",
   "metadata": {},
   "source": [
    "## Get info on each gene from mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ebd03d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-16000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 16001-17000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 17001-18000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 18001-19000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 19001-20000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 20001-21000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 21001-22000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 22001-23000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 23001-24000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 24001-25000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 25001-26000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 26001-27000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 27001-28000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 28001-29000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 29001-30000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 30001-31000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 31001-32000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 32001-33000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 33001-34000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 34001-35000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 35001-36000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 36001-37000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 37001-38000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 38001-39000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 39001-40000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 40001-41000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 41001-42000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 42001-43000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 43001-44000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 44001-45000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 45001-46000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 46001-47000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 47001-48000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 48001-49000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 49001-50000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 50001-51000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 51001-52000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 52001-53000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 53001-54000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 54001-55000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 55001-56000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 56001-57000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 57001-58000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 58001-59000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 59001-60000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 60001-61000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 61001-62000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 62001-63000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 63001-64000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 64001-65000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 65001-65009...\n",
      "INFO:biothings.client:done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR1</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210082</th>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR2</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000209082</th>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTL1</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  _id  _version      name symbol type_of_gene   alias  \\\n",
       "ensembl_gene_id                                                         \n",
       "ENSG00000210049  4558       2.0  tRNA-Phe   TRNF         tRNA     NaN   \n",
       "ENSG00000211459  4549       2.0    s-rRNA   RNR1         rRNA  MTRNR1   \n",
       "ENSG00000210077  4577       2.0  tRNA-Val   TRNV         tRNA    MTTV   \n",
       "ENSG00000210082  4550       2.0    l-rRNA   RNR2         rRNA  MTRNR2   \n",
       "ENSG00000209082  4567       2.0  tRNA-Leu  TRNL1         tRNA   MTTL1   \n",
       "\n",
       "                                                           summary notfound  \n",
       "ensembl_gene_id                                                              \n",
       "ENSG00000210049                                                NaN      NaN  \n",
       "ENSG00000211459  Enables DNA binding activity and DNA-binding t...      NaN  \n",
       "ENSG00000210077                                                NaN      NaN  \n",
       "ENSG00000210082  Enables G protein-coupled receptor binding act...      NaN  \n",
       "ENSG00000209082  Implicated in cardiomyopathy. [provided by All...      NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "mygene_output = mg.getgenes(\n",
    "    ensembl_ids_df[\"ensembl_gene_id\"],\n",
    "    fields=[\"symbol\", \"name\", \"summary\", \"type_of_gene\", \"alias\"],\n",
    "    as_dataframe=True,\n",
    ")\n",
    "\n",
    "mygene_output.index.rename(\"ensembl_gene_id\", inplace=True)\n",
    "mygene_output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23bb114e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations found for 63844 genes.\n",
      "No annotations found for 1176 genes.\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotations found for \" + str(sum(mygene_output[\"notfound\"].isna())) + \" genes.\")\n",
    "print(\n",
    "    \"No annotations found for \"\n",
    "    + str(sum(mygene_output[\"notfound\"] == True))\n",
    "    + \" genes.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf71470",
   "metadata": {},
   "source": [
    "# Part 2: Clean the data\n",
    "\n",
    "## Join and standardize columns / values\n",
    "\n",
    "For consistency with the `agora-data-tools` transform process, this uses the etl standardize functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "186d8cb8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65022, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR1</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>MTRNR2</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>MTTL1</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id chromosome_name hgnc_symbol   _id  _version      name  \\\n",
       "0  ENSG00000210049              MT       MT-TF  4558       2.0  tRNA-Phe   \n",
       "1  ENSG00000211459              MT     MT-RNR1  4549       2.0    s-rRNA   \n",
       "2  ENSG00000210077              MT       MT-TV  4577       2.0  tRNA-Val   \n",
       "3  ENSG00000210082              MT     MT-RNR2  4550       2.0    l-rRNA   \n",
       "4  ENSG00000209082              MT      MT-TL1  4567       2.0  tRNA-Leu   \n",
       "\n",
       "  symbol type_of_gene   alias  \\\n",
       "0   TRNF         tRNA     NaN   \n",
       "1   RNR1         rRNA  MTRNR1   \n",
       "2   TRNV         tRNA    MTTV   \n",
       "3   RNR2         rRNA  MTRNR2   \n",
       "4  TRNL1         tRNA   MTTL1   \n",
       "\n",
       "                                             summary notfound  \n",
       "0                                                NaN      NaN  \n",
       "1  Enables DNA binding activity and DNA-binding t...      NaN  \n",
       "2                                                NaN      NaN  \n",
       "3  Enables G protein-coupled receptor binding act...      NaN  \n",
       "4  Implicated in cardiomyopathy. [provided by All...      NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table_merged = pd.merge(\n",
    "    left=ensembl_ids_df,\n",
    "    right=mygene_output,\n",
    "    how=\"left\",\n",
    "    on=\"ensembl_gene_id\",\n",
    "    validate=\"many_to_many\",\n",
    ")\n",
    "\n",
    "gene_table_merged = utils.standardize_column_names(gene_table_merged)\n",
    "gene_table_merged = utils.standardize_values(gene_table_merged)\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9441115e",
   "metadata": {},
   "source": [
    "## Fix alias field\n",
    "\n",
    "Fix `NaN` values in the `alias` field and make sure every alias value is a list, not a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "285c10d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NaN or NULL alias values become empty lists\n",
    "for row in gene_table_merged.loc[gene_table_merged[\"alias\"].isnull(), \"alias\"].index:\n",
    "    gene_table_merged.at[row, \"alias\"] = []\n",
    "\n",
    "# Some alias values are a single string, not a list. Turn them into lists here.\n",
    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
    "    lambda cell: cell if isinstance(cell, list) else [cell]\n",
    ")\n",
    "\n",
    "# Some alias values are lists of lists or have duplicate values\n",
    "def flatten(row):\n",
    "    flattened = []\n",
    "    for item in row:\n",
    "        if isinstance(item, list):\n",
    "            flattened = flattened + item\n",
    "        else:\n",
    "            flattened.append(item)\n",
    "    return flattened\n",
    "\n",
    "\n",
    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
    "    lambda row: list(set(flatten(row)))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0be1b92",
   "metadata": {},
   "source": [
    "## Remove duplicate Ensembl IDs from the list. \n",
    "\n",
    "Duplicates in the list typically have the same Ensembl ID but different gene symbols. This usually happens when a single Ensembl ID maps to multiple Entrez IDs in the NCBI database. There's not a good way to reconcile this, so we first use entries whose `hgnc_symbol` from Biomart matches `symbol` from NCBI, then from the remaining duplicates just use the first entry in the list for each ensembl ID and discard the rest, which is what the Agora front end does. The gene symbols of duplicate rows are then added as aliases to the matching unique row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bc63cc53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4089</th>\n",
       "      <td>ENSG00000287838</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td>101927042</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC101927042</td>\n",
       "      <td>LOC101927042</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4090</th>\n",
       "      <td>ENSG00000287838</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td>124902157</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC124902157</td>\n",
       "      <td>LOC124902157</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5130</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>10</td>\n",
       "      <td>LINC00856</td>\n",
       "      <td>414243</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5131</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>10</td>\n",
       "      <td>LINC00856</td>\n",
       "      <td>414243</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5132</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>10</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>414243</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5133</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>10</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>414243</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8675</th>\n",
       "      <td>ENSG00000188660</td>\n",
       "      <td>21</td>\n",
       "      <td>LINC00319</td>\n",
       "      <td>124900467</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124900467</td>\n",
       "      <td>LOC124900467</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8676</th>\n",
       "      <td>ENSG00000188660</td>\n",
       "      <td>21</td>\n",
       "      <td>LINC00319</td>\n",
       "      <td>102724398</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized CH507-42P11.6</td>\n",
       "      <td>CH507-42P11.6</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12016</th>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>21</td>\n",
       "      <td></td>\n",
       "      <td>124905527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124905527</td>\n",
       "      <td>LOC124905527</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12017</th>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>21</td>\n",
       "      <td></td>\n",
       "      <td>124905312</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124905312</td>\n",
       "      <td>LOC124905312</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12018</th>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>21</td>\n",
       "      <td></td>\n",
       "      <td>124905468</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124905468</td>\n",
       "      <td>LOC124905468</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43337</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43338</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>285626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43519</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>15</td>\n",
       "      <td>GOLGA6L5P</td>\n",
       "      <td>100133220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>[GOLGA6L3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43520</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>15</td>\n",
       "      <td>GOLGA6L5P</td>\n",
       "      <td>642402</td>\n",
       "      <td>1.0</td>\n",
       "      <td>golgin A6 family like 17, pseudogene</td>\n",
       "      <td>GOLGA6L17P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>[GOLGA6L21P]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63071</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>101928626</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC101928626</td>\n",
       "      <td>LOC101928626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63072</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>124901156</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC124901156</td>\n",
       "      <td>LOC124901156</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63715</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966722</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63716</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966730</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966730</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63717</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966732</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966732</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63718</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966731</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966731</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63719</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966733</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966733</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id chromosome_name hgnc_symbol        _id  _version  \\\n",
       "4089   ENSG00000287838               9              101927042       1.0   \n",
       "4090   ENSG00000287838               9              124902157       2.0   \n",
       "5130   ENSG00000230417              10   LINC00856     414243       1.0   \n",
       "5131   ENSG00000230417              10   LINC00856     414243       1.0   \n",
       "5132   ENSG00000230417              10   LINC00595     414243       1.0   \n",
       "5133   ENSG00000230417              10   LINC00595     414243       1.0   \n",
       "8675   ENSG00000188660              21   LINC00319  124900467       1.0   \n",
       "8676   ENSG00000188660              21   LINC00319  102724398       1.0   \n",
       "12016  ENSG00000278903              21              124905527       1.0   \n",
       "12017  ENSG00000278903              21              124905312       1.0   \n",
       "12018  ENSG00000278903              21              124905468       1.0   \n",
       "43337  ENSG00000249738               5              105377683       1.0   \n",
       "43338  ENSG00000249738               5                 285626       1.0   \n",
       "43519  ENSG00000230373              15   GOLGA6L5P  100133220       1.0   \n",
       "43520  ENSG00000230373              15   GOLGA6L5P     642402       1.0   \n",
       "63071  ENSG00000293331               1              101928626       2.0   \n",
       "63072  ENSG00000293331               1              124901156       2.0   \n",
       "63715  ENSG00000276518                         NaN  128966722       2.0   \n",
       "63716  ENSG00000276518                         NaN  128966730       2.0   \n",
       "63717  ENSG00000276518                         NaN  128966732       2.0   \n",
       "63718  ENSG00000276518                         NaN  128966731       2.0   \n",
       "63719  ENSG00000276518                         NaN  128966733       2.0   \n",
       "\n",
       "                                                    name         symbol  \\\n",
       "4089                        uncharacterized LOC101927042   LOC101927042   \n",
       "4090                        uncharacterized LOC124902157   LOC124902157   \n",
       "5130          long intergenic non-protein coding RNA 595      LINC00595   \n",
       "5131          long intergenic non-protein coding RNA 595      LINC00595   \n",
       "5132          long intergenic non-protein coding RNA 595      LINC00595   \n",
       "5133          long intergenic non-protein coding RNA 595      LINC00595   \n",
       "8675                        uncharacterized LOC124900467   LOC124900467   \n",
       "8676                       uncharacterized CH507-42P11.6  CH507-42P11.6   \n",
       "12016                       uncharacterized LOC124905527   LOC124905527   \n",
       "12017                       uncharacterized LOC124905312   LOC124905312   \n",
       "12018                       uncharacterized LOC124905468   LOC124905468   \n",
       "43337                       uncharacterized LOC105377683   LOC105377683   \n",
       "43338                          uncharacterized LOC285626      LOC285626   \n",
       "43519                golgin A6 family like 3, pseudogene      GOLGA6L3P   \n",
       "43520               golgin A6 family like 17, pseudogene     GOLGA6L17P   \n",
       "63071                       uncharacterized LOC101928626   LOC101928626   \n",
       "63072                       uncharacterized LOC124901156   LOC124901156   \n",
       "63715  putative killer cell immunoglobulin-like recep...   LOC128966722   \n",
       "63716  putative killer cell immunoglobulin-like recep...   LOC128966730   \n",
       "63717  putative killer cell immunoglobulin-like recep...   LOC128966732   \n",
       "63718  putative killer cell immunoglobulin-like recep...   LOC128966731   \n",
       "63719  putative killer cell immunoglobulin-like recep...   LOC128966733   \n",
       "\n",
       "         type_of_gene         alias summary notfound  \n",
       "4089            ncRNA            []     NaN      NaN  \n",
       "4090            ncRNA            []     NaN      NaN  \n",
       "5130            ncRNA   [C10orf101]     NaN      NaN  \n",
       "5131            ncRNA   [C10orf101]     NaN      NaN  \n",
       "5132            ncRNA   [C10orf101]     NaN      NaN  \n",
       "5133            ncRNA   [C10orf101]     NaN      NaN  \n",
       "8675   protein-coding            []     NaN      NaN  \n",
       "8676            ncRNA            []     NaN      NaN  \n",
       "12016           ncRNA            []     NaN      NaN  \n",
       "12017           ncRNA            []     NaN      NaN  \n",
       "12018           ncRNA            []     NaN      NaN  \n",
       "43337           ncRNA            []     NaN      NaN  \n",
       "43338           ncRNA            []     NaN      NaN  \n",
       "43519          pseudo    [GOLGA6L3]     NaN      NaN  \n",
       "43520          pseudo  [GOLGA6L21P]     NaN      NaN  \n",
       "63071           ncRNA            []     NaN      NaN  \n",
       "63072           ncRNA            []     NaN      NaN  \n",
       "63715  protein-coding            []     NaN      NaN  \n",
       "63716  protein-coding            []     NaN      NaN  \n",
       "63717  protein-coding            []     NaN      NaN  \n",
       "63718  protein-coding            []     NaN      NaN  \n",
       "63719  protein-coding            []     NaN      NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated() will return true if the ID is a duplicate and is not the first one to appear the list.\n",
    "dupes = gene_table_merged[\"ensembl_gene_id\"].duplicated()\n",
    "dupe_vals = gene_table_merged[dupes]\n",
    "\n",
    "# Rows with duplicated Ensembl IDs\n",
    "all_duplicated = gene_table_merged.loc[\n",
    "    gene_table_merged[\"ensembl_gene_id\"].isin(dupe_vals[\"ensembl_gene_id\"])\n",
    "]\n",
    "all_duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "093a2e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 duplicated genes have been processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64998</th>\n",
       "      <td>ENSG00000277936</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>84311</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mitochondrial ribosomal protein L45</td>\n",
       "      <td>MRPL45</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[MRP-L45, L45mt, Mba1, mL45]</td>\n",
       "      <td>Mammalian mitochondrial ribosomal proteins are...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64999</th>\n",
       "      <td>ENSG00000277328</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65000</th>\n",
       "      <td>ENSG00000287838</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td>101927042</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC101927042</td>\n",
       "      <td>LOC101927042</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC124902157]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65001</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC285626]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65002</th>\n",
       "      <td>ENSG00000293331</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>101928626</td>\n",
       "      <td>2.0</td>\n",
       "      <td>uncharacterized LOC101928626</td>\n",
       "      <td>LOC101928626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC124901156]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65003</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>128966722</td>\n",
       "      <td>2.0</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC128966731, LOC128966733, LOC128966730, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65004</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>10</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>414243</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LINC00595, C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65005</th>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>21</td>\n",
       "      <td></td>\n",
       "      <td>124905527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124905527</td>\n",
       "      <td>LOC124905527</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC124905468, LOC124905312]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65006</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>15</td>\n",
       "      <td>GOLGA6L5P</td>\n",
       "      <td>100133220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>[GOLGA6L17P, GOLGA6L21P, GOLGA6L3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65007</th>\n",
       "      <td>ENSG00000188660</td>\n",
       "      <td>21</td>\n",
       "      <td>LINC00319</td>\n",
       "      <td>124900467</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124900467</td>\n",
       "      <td>LOC124900467</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[CH507-42P11.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id chromosome_name hgnc_symbol        _id _version  \\\n",
       "64998  ENSG00000277936                         NaN      84311      1.0   \n",
       "64999  ENSG00000277328                         NaN        NaN      NaN   \n",
       "65000  ENSG00000287838               9              101927042      1.0   \n",
       "65001  ENSG00000249738               5              105377683      1.0   \n",
       "65002  ENSG00000293331               1              101928626      2.0   \n",
       "65003  ENSG00000276518                         NaN  128966722      2.0   \n",
       "65004  ENSG00000230417              10   LINC00595     414243      1.0   \n",
       "65005  ENSG00000278903              21              124905527      1.0   \n",
       "65006  ENSG00000230373              15   GOLGA6L5P  100133220      1.0   \n",
       "65007  ENSG00000188660              21   LINC00319  124900467      1.0   \n",
       "\n",
       "                                                    name        symbol  \\\n",
       "64998                mitochondrial ribosomal protein L45        MRPL45   \n",
       "64999                                                NaN           NaN   \n",
       "65000                       uncharacterized LOC101927042  LOC101927042   \n",
       "65001                       uncharacterized LOC105377683  LOC105377683   \n",
       "65002                       uncharacterized LOC101928626  LOC101928626   \n",
       "65003  putative killer cell immunoglobulin-like recep...  LOC128966722   \n",
       "65004         long intergenic non-protein coding RNA 595     LINC00595   \n",
       "65005                       uncharacterized LOC124905527  LOC124905527   \n",
       "65006                golgin A6 family like 3, pseudogene     GOLGA6L3P   \n",
       "65007                       uncharacterized LOC124900467  LOC124900467   \n",
       "\n",
       "         type_of_gene                                              alias  \\\n",
       "64998  protein-coding                       [MRP-L45, L45mt, Mba1, mL45]   \n",
       "64999             NaN                                                 []   \n",
       "65000           ncRNA                                     [LOC124902157]   \n",
       "65001           ncRNA                                        [LOC285626]   \n",
       "65002           ncRNA                                     [LOC124901156]   \n",
       "65003  protein-coding  [LOC128966731, LOC128966733, LOC128966730, LOC...   \n",
       "65004           ncRNA                             [LINC00595, C10orf101]   \n",
       "65005           ncRNA                       [LOC124905468, LOC124905312]   \n",
       "65006          pseudo                 [GOLGA6L17P, GOLGA6L21P, GOLGA6L3]   \n",
       "65007  protein-coding                                    [CH507-42P11.6]   \n",
       "\n",
       "                                                 summary notfound  \n",
       "64998  Mammalian mitochondrial ribosomal proteins are...      NaN  \n",
       "64999                                                NaN     True  \n",
       "65000                                                NaN      NaN  \n",
       "65001                                                NaN      NaN  \n",
       "65002                                                NaN      NaN  \n",
       "65003                                                NaN      NaN  \n",
       "65004                                                NaN      NaN  \n",
       "65005                                                NaN      NaN  \n",
       "65006                                                NaN      NaN  \n",
       "65007                                                NaN      NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_dupes = set(gene_table_merged.index) - set(all_duplicated.index)\n",
    "keep_df = gene_table_merged.loc[list(non_dupes)].copy(deep=True)\n",
    "\n",
    "# For each duplicated Ensembl ID, collapse to 1 row and append that row to keep_df\n",
    "for ens_id in set(all_duplicated[\"ensembl_gene_id\"]):\n",
    "    group = all_duplicated.loc[all_duplicated[\"ensembl_gene_id\"] == ens_id].copy(\n",
    "        deep=True\n",
    "    )\n",
    "    matches = group[\"hgnc_symbol\"] == group[\"symbol\"]\n",
    "\n",
    "    # If there is a single entry with a matching symbol from NCBI, use that row and ignore the others\n",
    "    if sum(matches) == 1:\n",
    "        keep_df = keep_df.append(group.loc[matches], ignore_index=True)\n",
    "    # Multiple or no matching symbols, save the first entry in the list and add the other rows as aliases\n",
    "    else:\n",
    "        # For multiple matching symbols, discard non-matching entries and continue\n",
    "        if sum(matches) > 1:\n",
    "            group = group.loc[matches]\n",
    "\n",
    "        # Add all duplicate symbols and their aliases to the alias field of the first entry\n",
    "        for row in group.index[1:]:\n",
    "            group.at[group.index[0], \"alias\"].append(group[\"symbol\"][row])\n",
    "            if len(group.at[row, \"alias\"]) > 0:\n",
    "                group.at[group.index[0], \"alias\"] = (\n",
    "                    group.at[group.index[0], \"alias\"] + group[\"alias\"][row]\n",
    "                )\n",
    "\n",
    "        # Make sure we didn't add duplicate aliases\n",
    "        group.at[group.index[0], \"alias\"] = list(set(group.at[group.index[0], \"alias\"]))\n",
    "\n",
    "        # Keep the first row only, which now has all the aliases\n",
    "        keep_df = pd.concat([keep_df, group.iloc[0].to_frame().T], ignore_index=True)\n",
    "\n",
    "print(\n",
    "    str(len(all_duplicated.drop_duplicates(\"ensembl_gene_id\")))\n",
    "    + \" duplicated genes have been processed.\"\n",
    ")\n",
    "gene_table_merged = keep_df.reset_index(drop=True)\n",
    "gene_table_merged.tail(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "702c361d",
   "metadata": {},
   "source": [
    "# Part 3: Create Ensembl archive permalinks\n",
    "\n",
    "## Get a table of Ensembl archive URLs\n",
    "\n",
    "This is where we need to use the R biomaRt library specifically, instead of any of the available Python interfaces to Biomart, to get a table of Ensembl release versions and their corresponding archive URLs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a1bbdee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             name     date                                 url version\n",
      "1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n",
      "2     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111\n",
      "3     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110\n",
      "4     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n",
      "5     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n",
      "6     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n",
      "7     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n",
      "8     Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n",
      "9     Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n",
      "10    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n",
      "11    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n",
      "12    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n",
      "13    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n",
      "14     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n",
      "15     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n",
      "16     Ensembl 97 Jul 2019 https://jul2019.archive.ensembl.org      97\n",
      "17     Ensembl 96 Apr 2019 https://apr2019.archive.ensembl.org      96\n",
      "18     Ensembl 95 Jan 2019 https://jan2019.archive.ensembl.org      95\n",
      "19     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80\n",
      "20     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n",
      "21     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n",
      "22     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n",
      "   current_release\n",
      "1                 \n",
      "2                *\n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10                \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n",
      "17                \n",
      "18                \n",
      "19                \n",
      "20                \n",
      "21                \n",
      "22                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "archive_df = r.listEnsemblArchives()\n",
    "archive_df.to_csvfile(path=archive_filename, row_names=False, quote=False)\n",
    "\n",
    "print(archive_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e150d8e",
   "metadata": {},
   "source": [
    "## Query Ensembl for each gene's version\n",
    "\n",
    "Ensembl's REST API can only take 1000 genes at once, so this is looped to query groups of 1000. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9a747309",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "65008\n",
      "Querying genes 1 - 1000\n",
      "Querying genes 1001 - 2000\n",
      "Querying genes 2001 - 3000\n",
      "Querying genes 3001 - 4000\n",
      "Querying genes 4001 - 5000\n",
      "Querying genes 5001 - 6000\n",
      "Querying genes 6001 - 7000\n",
      "Querying genes 7001 - 8000\n",
      "Querying genes 8001 - 9000\n",
      "Querying genes 9001 - 10000\n",
      "Querying genes 10001 - 11000\n",
      "Querying genes 11001 - 12000\n",
      "Querying genes 12001 - 13000\n",
      "Querying genes 13001 - 14000\n",
      "Querying genes 14001 - 15000\n",
      "Querying genes 15001 - 16000\n",
      "Querying genes 16001 - 17000\n",
      "Querying genes 17001 - 18000\n",
      "Querying genes 18001 - 19000\n",
      "Querying genes 19001 - 20000\n",
      "Querying genes 20001 - 21000\n",
      "Querying genes 21001 - 22000\n",
      "Querying genes 22001 - 23000\n",
      "Querying genes 23001 - 24000\n",
      "Querying genes 24001 - 25000\n",
      "Querying genes 25001 - 26000\n",
      "Querying genes 26001 - 27000\n",
      "Querying genes 27001 - 28000\n",
      "Querying genes 28001 - 29000\n",
      "Querying genes 29001 - 30000\n",
      "Querying genes 30001 - 31000\n",
      "Querying genes 31001 - 32000\n",
      "Querying genes 32001 - 33000\n",
      "Querying genes 33001 - 34000\n",
      "Querying genes 34001 - 35000\n",
      "Querying genes 35001 - 36000\n",
      "Querying genes 36001 - 37000\n",
      "Querying genes 37001 - 38000\n",
      "Querying genes 38001 - 39000\n",
      "Querying genes 39001 - 40000\n",
      "Querying genes 40001 - 41000\n",
      "Querying genes 41001 - 42000\n",
      "Querying genes 42001 - 43000\n",
      "Querying genes 43001 - 44000\n",
      "Querying genes 44001 - 45000\n",
      "Querying genes 45001 - 46000\n",
      "Querying genes 46001 - 47000\n",
      "Querying genes 47001 - 48000\n",
      "Querying genes 48001 - 49000\n",
      "Querying genes 49001 - 50000\n",
      "Querying genes 50001 - 51000\n",
      "Querying genes 51001 - 52000\n",
      "Querying genes 52001 - 53000\n",
      "Querying genes 53001 - 54000\n",
      "Querying genes 54001 - 55000\n",
      "Querying genes 55001 - 56000\n",
      "Querying genes 56001 - 57000\n",
      "Querying genes 57001 - 58000\n",
      "Querying genes 58001 - 59000\n",
      "Querying genes 59001 - 60000\n",
      "Querying genes 60001 - 61000\n",
      "Querying genes 61001 - 62000\n",
      "Querying genes 62001 - 63000\n",
      "Querying genes 63001 - 64000\n",
      "Querying genes 64001 - 65000\n",
      "Querying genes 65001 - 65008\n",
      "65008\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>version</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>is_current</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>65003</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000276518.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65004</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000230417.12</td>\n",
       "      <td>[]</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65005</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000278903.5</td>\n",
       "      <td>[]</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65006</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000230373.9</td>\n",
       "      <td>[]</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65007</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000188660.5</td>\n",
       "      <td>[]</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000188660</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      assembly peptide release              latest possible_replacement  \\\n",
       "65003   GRCh38    None     111   ENSG00000276518.1                   []   \n",
       "65004   GRCh38    None     111  ENSG00000230417.12                   []   \n",
       "65005   GRCh38    None     111   ENSG00000278903.5                   []   \n",
       "65006   GRCh38    None     111   ENSG00000230373.9                   []   \n",
       "65007   GRCh38    None     111   ENSG00000188660.5                   []   \n",
       "\n",
       "       version               id  type is_current  \n",
       "65003        1  ENSG00000276518  Gene          1  \n",
       "65004       12  ENSG00000230417  Gene          1  \n",
       "65005        5  ENSG00000278903  Gene          1  \n",
       "65006        9  ENSG00000230373  Gene          1  \n",
       "65007        5  ENSG00000188660  Gene          1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://rest.ensembl.org/archive/id\"\n",
    "headers = {\"Content-Type\": \"application/json\", \"Accept\": \"application/json\"}\n",
    "\n",
    "ids = gene_table_merged[\"ensembl_gene_id\"].tolist()\n",
    "print(len(ids))\n",
    "\n",
    "# We can only query 1000 genes at a time\n",
    "batch_ind = range(0, len(ids), 1000)\n",
    "results = []\n",
    "\n",
    "for B in batch_ind:\n",
    "    end = min(len(ids), B + 1000)\n",
    "    print(\"Querying genes \" + str(B + 1) + \" - \" + str(end))\n",
    "\n",
    "    request_data = '{ \"id\" : ' + str(ids[B:end]) + \" }\"\n",
    "    request_data = request_data.replace(\"'\", '\"')\n",
    "\n",
    "    ok = False\n",
    "    tries = 0\n",
    "\n",
    "    while tries < 5 and not ok:\n",
    "        try:\n",
    "            res = requests.post(url, headers=headers, data=request_data)\n",
    "            ok = res.ok\n",
    "        except:\n",
    "            ok = False\n",
    "\n",
    "        tries = tries + 1\n",
    "\n",
    "        if not ok:\n",
    "            # res.raise_for_status()\n",
    "            print(\n",
    "                \"Error retrieving Ensembl versions for genes \"\n",
    "                + str(B + 1)\n",
    "                + \" - \"\n",
    "                + str(end)\n",
    "                + \". Trying again...\"\n",
    "            )\n",
    "        else:\n",
    "            results = results + res.json()\n",
    "            break\n",
    "\n",
    "print(len(results))\n",
    "\n",
    "versions = pd.json_normalize(results)\n",
    "\n",
    "versions.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5c108238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release\n",
       "100       22\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       34\n",
       "107       10\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111    63843\n",
       "80        21\n",
       "81         2\n",
       "82        10\n",
       "84       673\n",
       "87        61\n",
       "89        20\n",
       "91        75\n",
       "93        53\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         9\n",
       "99         7\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions.groupby(\"release\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bf5aecb1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "65008\n",
      "65008\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Check that all IDs are the same between the result and the gene table\n",
    "print(len(versions[\"id\"]))\n",
    "print(len(gene_table_merged))\n",
    "print(\n",
    "    all(versions[\"id\"].isin(gene_table_merged[\"ensembl_gene_id\"]))\n",
    "    and all(gene_table_merged[\"ensembl_gene_id\"].isin(versions[\"id\"]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7fc8bbcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure everything is GRCh38, not GRCh37\n",
    "all(versions[\"assembly\"] == \"GRCh38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e29252",
   "metadata": {},
   "source": [
    "## Create permalinks based on archive version\n",
    "\n",
    "**Not all of these versions have an archive.** We can go back to the closest previous archive for these but the link isn't guaranteed to work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0d5b5652",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "archive_table = pd.read_csv(archive_filename)\n",
    "\n",
    "# Remove GRCh37 from the archive list\n",
    "archive_table = archive_table[archive_table[\"version\"] != \"GRCh37\"].reset_index()\n",
    "\n",
    "archive_table[\"numeric_version\"] = archive_table[\"version\"].astype(int)\n",
    "\n",
    "\n",
    "def closest_release(release, archive_table):\n",
    "    if release in archive_table:\n",
    "        return release\n",
    "\n",
    "    return max([V for V in archive_table[\"numeric_version\"] if V <= release])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "337b2890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "closest_release\n",
       "80       915\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         9\n",
       "99         7\n",
       "100       22\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       34\n",
       "107       10\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111    63843\n",
       "dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[\"closest_release\"] = 0\n",
    "\n",
    "releases = versions[\"release\"].drop_duplicates().astype(int)\n",
    "\n",
    "# Only have to call closest_release once per version, instead of >70k times\n",
    "for release in releases:\n",
    "    versions.loc[\n",
    "        versions[\"release\"] == str(release), \"closest_release\"\n",
    "    ] = closest_release(release, archive_table)\n",
    "\n",
    "versions.groupby(\"closest_release\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "343e5006",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>version</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>111</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly peptide release             latest possible_replacement  version  \\\n",
       "0   GRCh38    None     111  ENSG00000210049.1                   []        1   \n",
       "1   GRCh38    None     111  ENSG00000211459.2                   []        2   \n",
       "2   GRCh38    None     111  ENSG00000210077.1                   []        1   \n",
       "3   GRCh38    None     111  ENSG00000210082.2                   []        2   \n",
       "4   GRCh38    None     111  ENSG00000209082.1                   []        1   \n",
       "\n",
       "                id  type is_current  closest_release  \\\n",
       "0  ENSG00000210049  Gene          1              111   \n",
       "1  ENSG00000211459  Gene          1              111   \n",
       "2  ENSG00000210077  Gene          1              111   \n",
       "3  ENSG00000210082  Gene          1              111   \n",
       "4  ENSG00000209082  Gene          1              111   \n",
       "\n",
       "                                           permalink  \n",
       "0  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4  https://jan2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[\"permalink\"] = \"\"\n",
    "\n",
    "for i in versions.index:\n",
    "    match = archive_table[\"numeric_version\"] == versions.at[i, \"closest_release\"]\n",
    "    url = archive_table.loc[match, \"url\"].to_string(index=False)\n",
    "    if len(url) > 0:\n",
    "        versions.at[i, \"permalink\"] = (\n",
    "            url + \"/Homo_sapiens/Gene/Summary?db=core;g=\" + versions.at[i, \"id\"]\n",
    "        )\n",
    "\n",
    "versions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4b01719d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>version</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63180</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000238909.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238909</td>\n",
       "      <td>Gene</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63181</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000265155.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000265155</td>\n",
       "      <td>Gene</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63183</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000275447.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000275447</td>\n",
       "      <td>Gene</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63184</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000263623.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000263623</td>\n",
       "      <td>Gene</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63190</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000238644.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238644</td>\n",
       "      <td>Gene</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      assembly peptide release             latest possible_replacement  \\\n",
       "63180   GRCh38    None      84  ENSG00000238909.1                   []   \n",
       "63181   GRCh38    None      84  ENSG00000265155.1                   []   \n",
       "63183   GRCh38    None      84  ENSG00000275447.1                   []   \n",
       "63184   GRCh38    None      84  ENSG00000263623.1                   []   \n",
       "63190   GRCh38    None      84  ENSG00000238644.1                   []   \n",
       "\n",
       "       version               id  type is_current  closest_release  \\\n",
       "63180        1  ENSG00000238909  Gene                          80   \n",
       "63181        1  ENSG00000265155  Gene                          80   \n",
       "63183        1  ENSG00000275447  Gene                          80   \n",
       "63184        1  ENSG00000263623  Gene                          80   \n",
       "63190        1  ENSG00000238644  Gene                          80   \n",
       "\n",
       "                                               permalink  \n",
       "63180  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "63181  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "63183  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "63184  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "63190  https://may2015.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[versions[\"closest_release\"] < 100].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c4128cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://jan2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000210049\n",
      "https://jan2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000210174\n"
     ]
    }
   ],
   "source": [
    "print(versions[\"permalink\"][0])\n",
    "print(versions[\"permalink\"][25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "73791e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Does every gene have an associated URL?\n",
    "url_base_len = len(archive_table[\"url\"][0]) + 1\n",
    "all([len(url) > url_base_len for url in versions[\"permalink\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b98b62a",
   "metadata": {},
   "source": [
    "# Part 4: Add permalinks to the gene table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f3edfd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65008, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>4558</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>4549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[MTRNR1]</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>4577</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[MTTV]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>4550</td>\n",
       "      <td>2.0</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[MTRNR2]</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>4567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>[MTTL1]</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id chromosome_name hgnc_symbol   _id _version      name  \\\n",
       "0  ENSG00000210049              MT       MT-TF  4558      2.0  tRNA-Phe   \n",
       "1  ENSG00000211459              MT     MT-RNR1  4549      2.0    s-rRNA   \n",
       "2  ENSG00000210077              MT       MT-TV  4577      2.0  tRNA-Val   \n",
       "3  ENSG00000210082              MT     MT-RNR2  4550      2.0    l-rRNA   \n",
       "4  ENSG00000209082              MT      MT-TL1  4567      2.0  tRNA-Leu   \n",
       "\n",
       "  symbol type_of_gene     alias  \\\n",
       "0   TRNF         tRNA        []   \n",
       "1   RNR1         rRNA  [MTRNR1]   \n",
       "2   TRNV         tRNA    [MTTV]   \n",
       "3   RNR2         rRNA  [MTRNR2]   \n",
       "4  TRNL1         tRNA   [MTTL1]   \n",
       "\n",
       "                                             summary notfound ensembl_release  \\\n",
       "0                                                NaN      NaN             111   \n",
       "1  Enables DNA binding activity and DNA-binding t...      NaN             111   \n",
       "2                                                NaN      NaN             111   \n",
       "3  Enables G protein-coupled receptor binding act...      NaN             111   \n",
       "4  Implicated in cardiomyopathy. [provided by All...      NaN             111   \n",
       "\n",
       "  possible_replacement                                          permalink  \n",
       "0                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4                   []  https://jan2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions = versions[[\"id\", \"release\", \"possible_replacement\", \"permalink\"]]\n",
    "versions.rename(\n",
    "    columns={\"id\": \"ensembl_gene_id\", \"release\": \"ensembl_release\"}, inplace=True\n",
    ")\n",
    "\n",
    "gene_table_merged = pd.merge(\n",
    "    left=gene_table_merged,\n",
    "    right=versions,\n",
    "    how=\"left\",\n",
    "    on=\"ensembl_gene_id\",\n",
    "    validate=\"one_to_one\",\n",
    ")\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01eba7c2",
   "metadata": {},
   "source": [
    "### Final cleanup\n",
    "Unfilled \"possible_replacement\" entries should be changed from NaN to empty lists. \n",
    "\n",
    "\"possible_replacement\" entries that have data in them exist as a list of dicts, and need to have the Ensembl IDs pulled out of them as a list of strings. \n",
    "\n",
    "Remove unneeded columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d0c07b7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>s-rRNA</td>\n",
       "      <td>[MTRNR1]</td>\n",
       "      <td>Enables DNA binding activity and DNA-binding t...</td>\n",
       "      <td>RNR1</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>[MTTV]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRNV</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>[MTRNR2]</td>\n",
       "      <td>Enables G protein-coupled receptor binding act...</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>[MTTL1]</td>\n",
       "      <td>Implicated in cardiomyopathy. [provided by All...</td>\n",
       "      <td>TRNL1</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65003</th>\n",
       "      <td>ENSG00000276518</td>\n",
       "      <td>putative killer cell immunoglobulin-like recep...</td>\n",
       "      <td>[LOC128966731, LOC128966733, LOC128966730, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC128966722</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65004</th>\n",
       "      <td>ENSG00000230417</td>\n",
       "      <td>long intergenic non-protein coding RNA 595</td>\n",
       "      <td>[LINC00595, C10orf101]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINC00595</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65005</th>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>uncharacterized LOC124905527</td>\n",
       "      <td>[LOC124905468, LOC124905312]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC124905527</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65006</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>[GOLGA6L17P, GOLGA6L21P, GOLGA6L3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65007</th>\n",
       "      <td>ENSG00000188660</td>\n",
       "      <td>uncharacterized LOC124900467</td>\n",
       "      <td>[CH507-42P11.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC124900467</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>111</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://jan2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65008 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id                                               name  \\\n",
       "0      ENSG00000210049                                           tRNA-Phe   \n",
       "1      ENSG00000211459                                             s-rRNA   \n",
       "2      ENSG00000210077                                           tRNA-Val   \n",
       "3      ENSG00000210082                                             l-rRNA   \n",
       "4      ENSG00000209082                                           tRNA-Leu   \n",
       "...                ...                                                ...   \n",
       "65003  ENSG00000276518  putative killer cell immunoglobulin-like recep...   \n",
       "65004  ENSG00000230417         long intergenic non-protein coding RNA 595   \n",
       "65005  ENSG00000278903                       uncharacterized LOC124905527   \n",
       "65006  ENSG00000230373                golgin A6 family like 3, pseudogene   \n",
       "65007  ENSG00000188660                       uncharacterized LOC124900467   \n",
       "\n",
       "                                                   alias  \\\n",
       "0                                                     []   \n",
       "1                                               [MTRNR1]   \n",
       "2                                                 [MTTV]   \n",
       "3                                               [MTRNR2]   \n",
       "4                                                [MTTL1]   \n",
       "...                                                  ...   \n",
       "65003  [LOC128966731, LOC128966733, LOC128966730, LOC...   \n",
       "65004                             [LINC00595, C10orf101]   \n",
       "65005                       [LOC124905468, LOC124905312]   \n",
       "65006                 [GOLGA6L17P, GOLGA6L21P, GOLGA6L3]   \n",
       "65007                                    [CH507-42P11.6]   \n",
       "\n",
       "                                                 summary        symbol  \\\n",
       "0                                                    NaN          TRNF   \n",
       "1      Enables DNA binding activity and DNA-binding t...          RNR1   \n",
       "2                                                    NaN          TRNV   \n",
       "3      Enables G protein-coupled receptor binding act...          RNR2   \n",
       "4      Implicated in cardiomyopathy. [provided by All...         TRNL1   \n",
       "...                                                  ...           ...   \n",
       "65003                                                NaN  LOC128966722   \n",
       "65004                                                NaN     LINC00595   \n",
       "65005                                                NaN  LOC124905527   \n",
       "65006                                                NaN     GOLGA6L3P   \n",
       "65007                                                NaN  LOC124900467   \n",
       "\n",
       "         type_of_gene ensembl_release possible_replacement  \\\n",
       "0                tRNA             111                   []   \n",
       "1                rRNA             111                   []   \n",
       "2                tRNA             111                   []   \n",
       "3                rRNA             111                   []   \n",
       "4                tRNA             111                   []   \n",
       "...               ...             ...                  ...   \n",
       "65003  protein-coding             111                   []   \n",
       "65004           ncRNA             111                   []   \n",
       "65005           ncRNA             111                   []   \n",
       "65006          pseudo             111                   []   \n",
       "65007  protein-coding             111                   []   \n",
       "\n",
       "                                               permalink  \n",
       "0      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "1      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "2      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "3      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "4      https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "...                                                  ...  \n",
       "65003  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "65004  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "65005  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "65006  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "65007  https://jan2024.archive.ensembl.org/Homo_sapie...  \n",
       "\n",
       "[65008 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for row in gene_table_merged.loc[\n",
    "    gene_table_merged[\"possible_replacement\"].isnull(), \"possible_replacement\"\n",
    "].index:\n",
    "    gene_table_merged.at[row, \"possible_replacement\"] = []\n",
    "\n",
    "gene_table_merged[\"possible_replacement\"] = gene_table_merged.apply(\n",
    "    lambda row: row[\"possible_replacement\"]\n",
    "    if len(row[\"possible_replacement\"]) == 0\n",
    "    else [x[\"stable_id\"] for x in row[\"possible_replacement\"]],\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "gene_table_merged = gene_table_merged[\n",
    "    [\n",
    "        \"ensembl_gene_id\",\n",
    "        \"name\",\n",
    "        \"alias\",\n",
    "        \"summary\",\n",
    "        \"symbol\",\n",
    "        \"type_of_gene\",\n",
    "        \"ensembl_release\",\n",
    "        \"possible_replacement\",\n",
    "        \"permalink\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "gene_table_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b371435",
   "metadata": {},
   "source": [
    "### Write to a file\n",
    "This will get uploaded to Synapse as [syn25953363](https://www.synapse.org/#!Synapse:syn25953363)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f2287922",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_merged = gene_table_merged.sort_values(by=\"ensembl_gene_id\").reset_index(\n",
    "    drop=True\n",
    ")\n",
    "gene_table_merged\n",
    "gene_table_merged.to_feather(\"../../output/gene_table_merged_GRCh38.p14.feather\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
