{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d9f808",
   "metadata": {},
   "source": [
    "# Preprocess Gene Annotations\n",
    "\n",
    "This notebook creates a table of gene annotations by:\n",
    "1. Querying Biomart for all Ensembl IDs in the database\n",
    "2. Querying MyGene for annotation about those IDs\n",
    "3. Querying Ensembl for the most recent Ensembl release for each ID\n",
    "4. Building a permalink to the Ensembl archive page for each ID\n",
    "\n",
    "This gene annotation table is read in by `agoradataprocessing/process.py` to be used in the `gene_info` transformation. \n",
    "\n",
    "***Note:*** *This notebook is exploratory and should eventually be converted to a Python script that is run through an automated process.*\n",
    "\n",
    "## Installation requirements\n",
    "\n",
    "#### Linux / Windows / Mac\n",
    "\n",
    "Install R: https://cran.r-project.org/\n",
    "\n",
    "Install Python and agora-data-tools following the instructions in this repository's README. This notebook assumes it is being run from the same `pipenv` virtual environment as agora-data-tools. \n",
    "\n",
    "Then install the following packages using `pip`:\n",
    "```\n",
    "pip install rpy2 mygene\n",
    "```\n",
    "\n",
    "#### Note for Macs with M1 chips (2020 and newer)\n",
    "\n",
    "Install as above, but make sure that your R installation is the arm64 version (R-4.X.X-arm64.pkg) so that the architecture matches what pip is using. \n",
    "You may also need to install an older version of `rpy2` on the Mac:\n",
    "```\n",
    "pip install rpy2==3.5.12\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2dc92c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from rpy2.robjects import r\n",
    "import pandas as pd\n",
    "import mygene\n",
    "import numpy as np\n",
    "import requests\n",
    "import agoradatatools.etl.utils as utils\n",
    "import agoradatatools.etl.extract as extract\n",
    "import preprocessing_utils\n",
    "\n",
    "r(\n",
    "    'if (!require(\"BiocManager\", character.only = TRUE)) { install.packages(\"BiocManager\") }'\n",
    ")\n",
    "r('if (!require(\"biomaRt\")) { BiocManager::install(\"biomaRt\") }')\n",
    "\n",
    "r.library(\"biomaRt\")\n",
    "\n",
    "ensembl_ids_filename = \"../../output/ensembl_id_list.txt\"\n",
    "archive_filename = \"../../output/ensembl_archive_list.csv\"\n",
    "config_filename = \"../../../../config.yaml\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec36f54c",
   "metadata": {},
   "source": [
    "# Part 1: Get gene annotation data\n",
    "\n",
    "## [Deprecated] Query Biomart for a list of all Ensembl IDs in the database of human genes. \n",
    "\n",
    "Here we use the R library `biomaRt`. There is no canonical Python library with the features we need for this script. \n",
    "\n",
    "*We no longer get all genes from BioMart, so this section is unused. The code is here in case we need it again.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e115d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "ensembl_ids_df = preprocessing_utils.r_query_biomart()\n",
    "ensembl_ids_df = preprocessing_utils.filter_hasgs(\n",
    "    df=ensembl_ids_df, chromosome_name_column=\"chromosome_name\"\n",
    ")\n",
    "print(str(ensembl_ids_df.shape[0]) + \" genes remaining after HASG filtering.\")\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f494e58f",
   "metadata": {},
   "source": [
    "## Get Ensembl IDs from data sets that will be processed by agora-data-tools\n",
    "\n",
    "Loop through all data sets in the config file to get all Ensembl IDs used in every data set. Exclude `gene_metadata` since that's the file we are building, and `druggability` since that data is deprecated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3fdbeec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.6.0) is available. Your version (4.0.0) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.6.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:synapseclient_default:Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 19 files:\n",
      "genes_biodomains:\tsyn44151254.5\n",
      "neuropath_regression_results:\tsyn22017882.5\n",
      "proteomics:\tsyn18689335.4\n",
      "proteomics_tmt:\tsyn35221005.2\n",
      "proteomics_srm:\tsyn52579640.4\n",
      "target_exp_validation_harmonized:\tsyn24184512.9\n",
      "metabolomics:\tsyn26064497.1\n",
      "gene_metadata:\tsyn25953363.13\n",
      "igap:\tsyn12514826.5\n",
      "eqtl:\tsyn12514912.3\n",
      "diff_exp_data:\tsyn27211942.1\n",
      "target_list:\tsyn12540368.51\n",
      "median_expression:\tsyn27211878.2\n",
      "druggability:\tsyn13363443.11\n",
      "tep_adi_info:\tsyn51942280.3\n",
      "team_info:\tsyn12615624.18\n",
      "team_member_info:\tsyn12615633.19\n",
      "overall_scores:\tsyn25575156.13\n",
      "networks:\tsyn11685347.1\n",
      "\n",
      "genes_biodomains has 591 NaN Ensembl IDs\n",
      "WARNING: no Ensembl ID column found for team_info!\n",
      "WARNING: no Ensembl ID column found for team_member_info!\n",
      "\n",
      "35858 Ensembl IDs found.\n",
      "['ENSG00000151650', 'ENSG00000168268', 'ENSG00000186310', 'ENSG00000204616', 'ENSG00000158467']\n"
     ]
    }
   ],
   "source": [
    "file_ensembl_list = preprocessing_utils.get_all_adt_ensembl_ids(\n",
    "    config_filename=config_filename,\n",
    "    exclude_files=[\"gene_metadata\", \"druggability\"],\n",
    "    token=None,\n",
    ")\n",
    "print(\"\")\n",
    "print(str(len(file_ensembl_list)) + \" Ensembl IDs found.\")\n",
    "print(file_ensembl_list[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fa76bfb",
   "metadata": {},
   "source": [
    "Create a data frame with these IDs so it can be merged with the MyGene query results below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1303e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35858\n"
     ]
    }
   ],
   "source": [
    "ensembl_ids_df = pd.DataFrame({\"ensembl_gene_id\": file_ensembl_list})\n",
    "\n",
    "\"\"\" Removed due to no longer getting genes from BioMart, but saving code\n",
    "# Add Ensembl IDs that are in the files but not in the biomart result\n",
    "missing = set(file_ensembl_list) - set(ensembl_ids_df[\"ensembl_gene_id\"])\n",
    "print(\n",
    "    str(len(missing))\n",
    "    + \" genes from the data files are missing from Biomart results and will be added.\"\n",
    ")\n",
    "\n",
    "missing_df = pd.DataFrame({\"ensembl_gene_id\": list(missing), \"chromosome_name\": \"\"})\n",
    "ensembl_ids_df = pd.concat([ensembl_ids_df, missing_df])\n",
    "\"\"\"\n",
    "\n",
    "ensembl_ids_df = ensembl_ids_df.dropna(subset=[\"ensembl_gene_id\"])\n",
    "print(len(ensembl_ids_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4e7a37c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to a file to save the list of IDs\n",
    "ensembl_ids_df.to_csv(\n",
    "    path_or_buf=ensembl_ids_filename, sep=\"\\t\", header=False, index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13ef6ec",
   "metadata": {},
   "source": [
    "## Get info on each gene from mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7ebd03d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-16000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 16001-17000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 17001-18000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 18001-19000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 19001-20000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 20001-21000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 21001-22000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 22001-23000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 23001-24000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 24001-25000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 25001-26000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 26001-27000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 27001-28000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 28001-29000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 29001-30000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 30001-31000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 31001-32000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 32001-33000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 33001-34000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 34001-35000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 35001-35858...\n",
      "INFO:biothings.client:done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>alias</th>\n",
       "      <th>name</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000151650</th>\n",
       "      <td>27287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[HPX42B, NA88A, VENTX2]</td>\n",
       "      <td>VENT homeobox</td>\n",
       "      <td>This gene encodes a member of the Vent family ...</td>\n",
       "      <td>VENTX</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000168268</th>\n",
       "      <td>64943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'-nucleotidase domain containing 2</td>\n",
       "      <td>Predicted to enable 5'-nucleotidase activity. ...</td>\n",
       "      <td>NT5DC2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186310</th>\n",
       "      <td>4675</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MB20, NPL3]</td>\n",
       "      <td>nucleosome assembly protein 1 like 3</td>\n",
       "      <td>This gene is intronless and encodes a member o...</td>\n",
       "      <td>NAP1L3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000204616</th>\n",
       "      <td>11074</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[C6orf13, HCG1, HCGI, RNF]</td>\n",
       "      <td>tripartite motif containing 31</td>\n",
       "      <td>This gene encodes a protein that functions as ...</td>\n",
       "      <td>TRIM31</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158467</th>\n",
       "      <td>23382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[ADOHCYASE3, IRBIT2]</td>\n",
       "      <td>adenosylhomocysteinase like 2</td>\n",
       "      <td>The protein encoded by this gene acts as a hom...</td>\n",
       "      <td>AHCYL2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   _id  _version                       alias  \\\n",
       "ensembl_gene_id                                                \n",
       "ENSG00000151650  27287       1.0     [HPX42B, NA88A, VENTX2]   \n",
       "ENSG00000168268  64943       1.0                         NaN   \n",
       "ENSG00000186310   4675       1.0                [MB20, NPL3]   \n",
       "ENSG00000204616  11074       1.0  [C6orf13, HCG1, HCGI, RNF]   \n",
       "ENSG00000158467  23382       1.0        [ADOHCYASE3, IRBIT2]   \n",
       "\n",
       "                                                 name  \\\n",
       "ensembl_gene_id                                         \n",
       "ENSG00000151650                         VENT homeobox   \n",
       "ENSG00000168268   5'-nucleotidase domain containing 2   \n",
       "ENSG00000186310  nucleosome assembly protein 1 like 3   \n",
       "ENSG00000204616        tripartite motif containing 31   \n",
       "ENSG00000158467         adenosylhomocysteinase like 2   \n",
       "\n",
       "                                                           summary  symbol  \\\n",
       "ensembl_gene_id                                                              \n",
       "ENSG00000151650  This gene encodes a member of the Vent family ...   VENTX   \n",
       "ENSG00000168268  Predicted to enable 5'-nucleotidase activity. ...  NT5DC2   \n",
       "ENSG00000186310  This gene is intronless and encodes a member o...  NAP1L3   \n",
       "ENSG00000204616  This gene encodes a protein that functions as ...  TRIM31   \n",
       "ENSG00000158467  The protein encoded by this gene acts as a hom...  AHCYL2   \n",
       "\n",
       "                   type_of_gene notfound  \n",
       "ensembl_gene_id                           \n",
       "ENSG00000151650  protein-coding      NaN  \n",
       "ENSG00000168268  protein-coding      NaN  \n",
       "ENSG00000186310  protein-coding      NaN  \n",
       "ENSG00000204616  protein-coding      NaN  \n",
       "ENSG00000158467  protein-coding      NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "mygene_output = mg.getgenes(\n",
    "    ensembl_ids_df[\"ensembl_gene_id\"],\n",
    "    fields=[\"symbol\", \"name\", \"summary\", \"type_of_gene\", \"alias\"],\n",
    "    as_dataframe=True,\n",
    ")\n",
    "\n",
    "mygene_output.index.rename(\"ensembl_gene_id\", inplace=True)\n",
    "mygene_output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "23bb114e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations found for 34655 genes.\n",
      "No annotations found for 1206 genes.\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotations found for \" + str(sum(mygene_output[\"notfound\"].isna())) + \" genes.\")\n",
    "print(\n",
    "    \"No annotations found for \"\n",
    "    + str(sum(mygene_output[\"notfound\"] == True))\n",
    "    + \" genes.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf71470",
   "metadata": {},
   "source": [
    "# Part 2: Clean the data\n",
    "\n",
    "## Join and standardize columns / values\n",
    "\n",
    "For consistency with the `agora-data-tools` transform process, this uses the etl standardize functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "186d8cb8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35861, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>alias</th>\n",
       "      <th>name</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000151650</td>\n",
       "      <td>27287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[HPX42B, NA88A, VENTX2]</td>\n",
       "      <td>VENT homeobox</td>\n",
       "      <td>This gene encodes a member of the Vent family ...</td>\n",
       "      <td>VENTX</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000168268</td>\n",
       "      <td>64943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'-nucleotidase domain containing 2</td>\n",
       "      <td>Predicted to enable 5'-nucleotidase activity. ...</td>\n",
       "      <td>NT5DC2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186310</td>\n",
       "      <td>4675</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MB20, NPL3]</td>\n",
       "      <td>nucleosome assembly protein 1 like 3</td>\n",
       "      <td>This gene is intronless and encodes a member o...</td>\n",
       "      <td>NAP1L3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000204616</td>\n",
       "      <td>11074</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[C6orf13, HCG1, HCGI, RNF]</td>\n",
       "      <td>tripartite motif containing 31</td>\n",
       "      <td>This gene encodes a protein that functions as ...</td>\n",
       "      <td>TRIM31</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000158467</td>\n",
       "      <td>23382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[ADOHCYASE3, IRBIT2]</td>\n",
       "      <td>adenosylhomocysteinase like 2</td>\n",
       "      <td>The protein encoded by this gene acts as a hom...</td>\n",
       "      <td>AHCYL2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id    _id  _version                       alias  \\\n",
       "0  ENSG00000151650  27287       1.0     [HPX42B, NA88A, VENTX2]   \n",
       "1  ENSG00000168268  64943       1.0                         NaN   \n",
       "2  ENSG00000186310   4675       1.0                [MB20, NPL3]   \n",
       "3  ENSG00000204616  11074       1.0  [C6orf13, HCG1, HCGI, RNF]   \n",
       "4  ENSG00000158467  23382       1.0        [ADOHCYASE3, IRBIT2]   \n",
       "\n",
       "                                   name  \\\n",
       "0                         VENT homeobox   \n",
       "1   5'-nucleotidase domain containing 2   \n",
       "2  nucleosome assembly protein 1 like 3   \n",
       "3        tripartite motif containing 31   \n",
       "4         adenosylhomocysteinase like 2   \n",
       "\n",
       "                                             summary  symbol    type_of_gene  \\\n",
       "0  This gene encodes a member of the Vent family ...   VENTX  protein-coding   \n",
       "1  Predicted to enable 5'-nucleotidase activity. ...  NT5DC2  protein-coding   \n",
       "2  This gene is intronless and encodes a member o...  NAP1L3  protein-coding   \n",
       "3  This gene encodes a protein that functions as ...  TRIM31  protein-coding   \n",
       "4  The protein encoded by this gene acts as a hom...  AHCYL2  protein-coding   \n",
       "\n",
       "  notfound  \n",
       "0      NaN  \n",
       "1      NaN  \n",
       "2      NaN  \n",
       "3      NaN  \n",
       "4      NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table_merged = pd.merge(\n",
    "    left=ensembl_ids_df,\n",
    "    right=mygene_output,\n",
    "    how=\"left\",\n",
    "    on=\"ensembl_gene_id\",\n",
    "    validate=\"many_to_many\",\n",
    ")\n",
    "\n",
    "gene_table_merged = utils.standardize_column_names(gene_table_merged)\n",
    "gene_table_merged = utils.standardize_values(gene_table_merged)\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9441115e",
   "metadata": {},
   "source": [
    "## Fix alias field\n",
    "\n",
    "Fix `NaN` values in the `alias` field and make sure every alias value is a list, not a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "285c10d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NaN or NULL alias values become empty lists\n",
    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
    "    lambda cell: cell if cell is not np.NaN else []\n",
    ")\n",
    "\n",
    "# Some alias values are a single string, not a list. Turn them into lists here.\n",
    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
    "    lambda cell: cell if isinstance(cell, list) else [cell]\n",
    ")\n",
    "\n",
    "\n",
    "# Some alias values are lists of lists or have duplicate values\n",
    "def flatten(row):\n",
    "    flattened = []\n",
    "    for item in row:\n",
    "        if isinstance(item, list):\n",
    "            flattened = flattened + item\n",
    "        else:\n",
    "            flattened.append(item)\n",
    "    return flattened\n",
    "\n",
    "\n",
    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
    "    lambda row: list(set(flatten(row)))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0be1b92",
   "metadata": {},
   "source": [
    "## Remove duplicate Ensembl IDs from the list. \n",
    "\n",
    "Duplicates in the list typically have the same Ensembl ID but different gene symbols. This usually happens when a single Ensembl ID maps to multiple Entrez IDs in the NCBI database. There's not a good way to reconcile this, so we first check for entries whose `symbol` is something other than \"LOC#######\", and designate that entry as the main row. If there are multiple or zero entries meeting that criteria, we just use the first entry in the list for each ensembl ID and discard the rest, which is what the Agora front end does. The gene symbols of duplicate rows are then added as aliases to the matching unique row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bc63cc53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>alias</th>\n",
       "      <th>name</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19626</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>285626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19627</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24698</th>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>3802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[CD158A, NKAT1, KIR2DL3, KIR-K64, NKAT-1, p58....</td>\n",
       "      <td>killer cell immunoglobulin like receptor, two ...</td>\n",
       "      <td>Killer cell immunoglobulin-like receptors (KIR...</td>\n",
       "      <td>KIR2DL1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24699</th>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>124900571</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>killer cell immunoglobulin-like receptor 2DS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC124900571</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29514</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>100133220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[GOLGA6L3]</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29515</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>642402</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[GOLGA6L21P]</td>\n",
       "      <td>golgin A6 family like 17, pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOLGA6L17P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id        _id  _version  \\\n",
       "19626  ENSG00000249738     285626       1.0   \n",
       "19627  ENSG00000249738  105377683       1.0   \n",
       "24698  ENSG00000276387       3802       1.0   \n",
       "24699  ENSG00000276387  124900571       1.0   \n",
       "29514  ENSG00000230373  100133220       1.0   \n",
       "29515  ENSG00000230373     642402       1.0   \n",
       "\n",
       "                                                   alias  \\\n",
       "19626                                                 []   \n",
       "19627                                                 []   \n",
       "24698  [CD158A, NKAT1, KIR2DL3, KIR-K64, NKAT-1, p58....   \n",
       "24699                                                 []   \n",
       "29514                                         [GOLGA6L3]   \n",
       "29515                                       [GOLGA6L21P]   \n",
       "\n",
       "                                                    name  \\\n",
       "19626                          uncharacterized LOC285626   \n",
       "19627                       uncharacterized LOC105377683   \n",
       "24698  killer cell immunoglobulin like receptor, two ...   \n",
       "24699      killer cell immunoglobulin-like receptor 2DS1   \n",
       "29514                golgin A6 family like 3, pseudogene   \n",
       "29515               golgin A6 family like 17, pseudogene   \n",
       "\n",
       "                                                 summary        symbol  \\\n",
       "19626                                                NaN     LOC285626   \n",
       "19627                                                NaN  LOC105377683   \n",
       "24698  Killer cell immunoglobulin-like receptors (KIR...       KIR2DL1   \n",
       "24699                                                NaN  LOC124900571   \n",
       "29514                                                NaN     GOLGA6L3P   \n",
       "29515                                                NaN    GOLGA6L17P   \n",
       "\n",
       "         type_of_gene notfound  \n",
       "19626           ncRNA      NaN  \n",
       "19627           ncRNA      NaN  \n",
       "24698  protein-coding      NaN  \n",
       "24699  protein-coding      NaN  \n",
       "29514          pseudo      NaN  \n",
       "29515          pseudo      NaN  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated() will return true if the ID is a duplicate and is not the first one to appear the list.\n",
    "dupes = gene_table_merged[\"ensembl_gene_id\"].duplicated()\n",
    "dupe_vals = gene_table_merged[dupes]\n",
    "\n",
    "# Rows with duplicated Ensembl IDs\n",
    "all_duplicated = gene_table_merged.loc[\n",
    "    gene_table_merged[\"ensembl_gene_id\"].isin(dupe_vals[\"ensembl_gene_id\"])\n",
    "]\n",
    "all_duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "093a2e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 duplicated genes have been processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>alias</th>\n",
       "      <th>name</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35848</th>\n",
       "      <td>ENSG00000085998</td>\n",
       "      <td>55624</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[RP76, LGMDR15, LGMD2O, gnT-I.2, GNTI.2, GnT I...</td>\n",
       "      <td>protein O-linked mannose N-acetylglucosaminylt...</td>\n",
       "      <td>This gene encodes a type II transmembrane prot...</td>\n",
       "      <td>POMGNT1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35849</th>\n",
       "      <td>ENSG00000285081</td>\n",
       "      <td>ENSG00000285081</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35850</th>\n",
       "      <td>ENSG00000126822</td>\n",
       "      <td>26030</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[ARHGEF43, KIAA0599]</td>\n",
       "      <td>pleckstrin homology and RhoGEF domain containi...</td>\n",
       "      <td>Predicted to enable guanyl-nucleotide exchange...</td>\n",
       "      <td>PLEKHG3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35851</th>\n",
       "      <td>ENSG00000187240</td>\n",
       "      <td>79659</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[DHC2, hdhc11, DNCH2, SRTD3, SRPS2B, ATD3, DHC...</td>\n",
       "      <td>dynein cytoplasmic 2 heavy chain 1</td>\n",
       "      <td>This gene encodes a large cytoplasmic dynein p...</td>\n",
       "      <td>DYNC2H1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35852</th>\n",
       "      <td>ENSG00000101470</td>\n",
       "      <td>7125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[CMYP15, CMYO15, CFAP85, FAP85, MYONRI]</td>\n",
       "      <td>troponin C2, fast skeletal type</td>\n",
       "      <td>Troponin (Tn), a key protein complex in the re...</td>\n",
       "      <td>TNNC2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35853</th>\n",
       "      <td>ENSG00000241472</td>\n",
       "      <td>100506994</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>PTPRG antisense RNA 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTPRG-AS1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35854</th>\n",
       "      <td>ENSG00000133106</td>\n",
       "      <td>94240</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[BRESI1]</td>\n",
       "      <td>epithelial stromal interaction 1</td>\n",
       "      <td>The protein encoded by this gene has been show...</td>\n",
       "      <td>EPSTI1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35855</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>100133220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[GOLGA6L3, GOLGA6L21P, GOLGA6L17P]</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35856</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>285626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[LOC105377683]</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35857</th>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>3802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[CD158A, NKAT1, KIR2DL3, LOC124900571, KIR-K64...</td>\n",
       "      <td>killer cell immunoglobulin like receptor, two ...</td>\n",
       "      <td>Killer cell immunoglobulin-like receptors (KIR...</td>\n",
       "      <td>KIR2DL1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id              _id _version  \\\n",
       "35848  ENSG00000085998            55624      1.0   \n",
       "35849  ENSG00000285081  ENSG00000285081      1.0   \n",
       "35850  ENSG00000126822            26030      1.0   \n",
       "35851  ENSG00000187240            79659      1.0   \n",
       "35852  ENSG00000101470             7125      1.0   \n",
       "35853  ENSG00000241472        100506994      1.0   \n",
       "35854  ENSG00000133106            94240      1.0   \n",
       "35855  ENSG00000230373        100133220      1.0   \n",
       "35856  ENSG00000249738           285626      1.0   \n",
       "35857  ENSG00000276387             3802      1.0   \n",
       "\n",
       "                                                   alias  \\\n",
       "35848  [RP76, LGMDR15, LGMD2O, gnT-I.2, GNTI.2, GnT I...   \n",
       "35849                                                 []   \n",
       "35850                               [ARHGEF43, KIAA0599]   \n",
       "35851  [DHC2, hdhc11, DNCH2, SRTD3, SRPS2B, ATD3, DHC...   \n",
       "35852            [CMYP15, CMYO15, CFAP85, FAP85, MYONRI]   \n",
       "35853                                                 []   \n",
       "35854                                           [BRESI1]   \n",
       "35855                 [GOLGA6L3, GOLGA6L21P, GOLGA6L17P]   \n",
       "35856                                     [LOC105377683]   \n",
       "35857  [CD158A, NKAT1, KIR2DL3, LOC124900571, KIR-K64...   \n",
       "\n",
       "                                                    name  \\\n",
       "35848  protein O-linked mannose N-acetylglucosaminylt...   \n",
       "35849                                                NaN   \n",
       "35850  pleckstrin homology and RhoGEF domain containi...   \n",
       "35851                 dynein cytoplasmic 2 heavy chain 1   \n",
       "35852                    troponin C2, fast skeletal type   \n",
       "35853                              PTPRG antisense RNA 1   \n",
       "35854                   epithelial stromal interaction 1   \n",
       "35855                golgin A6 family like 3, pseudogene   \n",
       "35856                          uncharacterized LOC285626   \n",
       "35857  killer cell immunoglobulin like receptor, two ...   \n",
       "\n",
       "                                                 summary     symbol  \\\n",
       "35848  This gene encodes a type II transmembrane prot...    POMGNT1   \n",
       "35849                                                NaN        NaN   \n",
       "35850  Predicted to enable guanyl-nucleotide exchange...    PLEKHG3   \n",
       "35851  This gene encodes a large cytoplasmic dynein p...    DYNC2H1   \n",
       "35852  Troponin (Tn), a key protein complex in the re...      TNNC2   \n",
       "35853                                                NaN  PTPRG-AS1   \n",
       "35854  The protein encoded by this gene has been show...     EPSTI1   \n",
       "35855                                                NaN  GOLGA6L3P   \n",
       "35856                                                NaN  LOC285626   \n",
       "35857  Killer cell immunoglobulin-like receptors (KIR...    KIR2DL1   \n",
       "\n",
       "         type_of_gene notfound  \n",
       "35848  protein-coding      NaN  \n",
       "35849             NaN      NaN  \n",
       "35850  protein-coding      NaN  \n",
       "35851  protein-coding      NaN  \n",
       "35852  protein-coding      NaN  \n",
       "35853           ncRNA      NaN  \n",
       "35854  protein-coding      NaN  \n",
       "35855          pseudo      NaN  \n",
       "35856           ncRNA      NaN  \n",
       "35857  protein-coding      NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_df = gene_table_merged.drop(all_duplicated.index)\n",
    "\n",
    "# For each duplicated Ensembl ID, collapse to 1 row and append that row to keep_df\n",
    "for ens_id in set(all_duplicated[\"ensembl_gene_id\"]):\n",
    "    group = all_duplicated.loc[all_duplicated[\"ensembl_gene_id\"] == ens_id].copy(\n",
    "        deep=True\n",
    "    )\n",
    "    # Put any entries with symbols that aren't \"LOC#####\" at the top of the data frame\n",
    "    matches = group[\"symbol\"].str.startswith(\"LOC\") == False\n",
    "    group = pd.concat([group.loc[matches], group.loc[matches == False]]).reset_index(\n",
    "        drop=True\n",
    "    )\n",
    "\n",
    "    # Add all duplicate symbols and their aliases to the alias field of the first entry\n",
    "    for row in group.index[1:]:\n",
    "        group.at[group.index[0], \"alias\"].append(group[\"symbol\"][row])\n",
    "        if len(group.at[row, \"alias\"]) > 0:\n",
    "            group.at[group.index[0], \"alias\"] = (\n",
    "                group.at[group.index[0], \"alias\"] + group[\"alias\"][row]\n",
    "            )\n",
    "\n",
    "    # Make sure we didn't add duplicate aliases\n",
    "    group.at[group.index[0], \"alias\"] = list(set(group.at[group.index[0], \"alias\"]))\n",
    "\n",
    "    # Keep the first row only, which now has all the aliases\n",
    "    keep_df = pd.concat([keep_df, group.iloc[0].to_frame().T], ignore_index=True)\n",
    "\n",
    "print(\n",
    "    str(len(all_duplicated.drop_duplicates(\"ensembl_gene_id\")))\n",
    "    + \" duplicated genes have been processed.\"\n",
    ")\n",
    "gene_table_merged = keep_df.reset_index(drop=True)\n",
    "gene_table_merged.tail(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "702c361d",
   "metadata": {},
   "source": [
    "# Part 3: Create Ensembl archive permalinks\n",
    "\n",
    "## Get a table of Ensembl archive URLs\n",
    "\n",
    "This is where we need to use the R biomaRt library specifically, instead of any of the available Python interfaces to Biomart, to get a table of Ensembl release versions and their corresponding archive URLs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4a1bbdee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             name     date                                 url version\n",
      "1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n",
      "2     Ensembl 113 Oct 2024 https://oct2024.archive.ensembl.org     113\n",
      "3     Ensembl 112 May 2024 https://may2024.archive.ensembl.org     112\n",
      "4     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111\n",
      "5     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110\n",
      "6     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n",
      "7     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n",
      "8     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n",
      "9     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n",
      "10    Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n",
      "11    Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n",
      "12    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n",
      "13    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n",
      "14    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n",
      "15    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n",
      "16     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n",
      "17     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n",
      "18     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80\n",
      "19     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n",
      "20     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n",
      "21     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n",
      "   current_release\n",
      "1                 \n",
      "2                *\n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10                \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n",
      "17                \n",
      "18                \n",
      "19                \n",
      "20                \n",
      "21                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "archive_df = r.listEnsemblArchives()\n",
    "archive_df.to_csvfile(path=archive_filename, row_names=False, quote=False)\n",
    "\n",
    "print(archive_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e150d8e",
   "metadata": {},
   "source": [
    "## Query Ensembl for each gene's version\n",
    "\n",
    "Ensembl's REST API can only take 1000 genes at once, so this is looped to query groups of 1000. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9a747309",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying genes 1 - 1000\n",
      "Querying genes 1001 - 2000\n",
      "Querying genes 2001 - 3000\n",
      "Querying genes 3001 - 4000\n",
      "Querying genes 4001 - 5000\n",
      "Querying genes 5001 - 6000\n",
      "Querying genes 6001 - 7000\n",
      "Querying genes 7001 - 8000\n",
      "Querying genes 8001 - 9000\n",
      "Querying genes 9001 - 10000\n",
      "Querying genes 10001 - 11000\n",
      "Querying genes 11001 - 12000\n",
      "Querying genes 12001 - 13000\n",
      "Querying genes 13001 - 14000\n",
      "Querying genes 14001 - 15000\n",
      "Querying genes 15001 - 16000\n",
      "Querying genes 16001 - 17000\n",
      "Querying genes 17001 - 18000\n",
      "Querying genes 18001 - 19000\n",
      "Querying genes 19001 - 20000\n",
      "Querying genes 20001 - 21000\n",
      "Querying genes 21001 - 22000\n",
      "Querying genes 22001 - 23000\n",
      "Querying genes 23001 - 24000\n",
      "Querying genes 24001 - 25000\n",
      "Querying genes 25001 - 26000\n",
      "Querying genes 26001 - 27000\n",
      "Querying genes 27001 - 28000\n",
      "Querying genes 28001 - 29000\n",
      "Querying genes 29001 - 30000\n",
      "Querying genes 30001 - 31000\n",
      "Querying genes 31001 - 32000\n",
      "Querying genes 32001 - 33000\n",
      "Querying genes 33001 - 34000\n",
      "Querying genes 34001 - 35000\n",
      "Querying genes 35001 - 35858\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th>is_current</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35853</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000241472.9</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000241472</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35854</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000133106.15</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000133106</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35855</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000230373.9</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35856</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000249738.11</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35857</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000276387.4</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      assembly peptide possible_replacement release              latest  type  \\\n",
       "35853   GRCh38    None                   []     113   ENSG00000241472.9  Gene   \n",
       "35854   GRCh38    None                   []     113  ENSG00000133106.15  Gene   \n",
       "35855   GRCh38    None                   []     113   ENSG00000230373.9  Gene   \n",
       "35856   GRCh38    None                   []     113  ENSG00000249738.11  Gene   \n",
       "35857   GRCh38    None                   []     113   ENSG00000276387.4  Gene   \n",
       "\n",
       "                    id  version is_current  \n",
       "35853  ENSG00000241472        9          1  \n",
       "35854  ENSG00000133106       15          1  \n",
       "35855  ENSG00000230373        9          1  \n",
       "35856  ENSG00000249738       11          1  \n",
       "35857  ENSG00000276387        4          1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions = preprocessing_utils.query_ensembl_version_api(\n",
    "    ensembl_ids=gene_table_merged[\"ensembl_gene_id\"].tolist()\n",
    ")\n",
    "\n",
    "versions.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5c108238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release\n",
       "100       21\n",
       "101        8\n",
       "102       16\n",
       "103       12\n",
       "104       17\n",
       "105       10\n",
       "106       35\n",
       "107       12\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111       52\n",
       "112      354\n",
       "113    34303\n",
       "80        21\n",
       "81         2\n",
       "82        10\n",
       "84       673\n",
       "87        61\n",
       "89        20\n",
       "91        67\n",
       "93        50\n",
       "95        33\n",
       "96        31\n",
       "97        17\n",
       "98         9\n",
       "99         5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions.groupby(\"release\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bf5aecb1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35858\n",
      "35858\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Check that all IDs are the same between the result and the gene table\n",
    "print(len(versions[\"id\"]))\n",
    "print(len(gene_table_merged))\n",
    "print(\n",
    "    all(versions[\"id\"].isin(gene_table_merged[\"ensembl_gene_id\"]))\n",
    "    and all(gene_table_merged[\"ensembl_gene_id\"].isin(versions[\"id\"]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7fc8bbcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure everything is GRCh38, not GRCh37\n",
    "all(versions[\"assembly\"] == \"GRCh38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e29252",
   "metadata": {},
   "source": [
    "## Create permalinks based on archive version\n",
    "\n",
    "**Not all of these versions have an archive.** We can go back to the closest previous archive for these but the link isn't guaranteed to work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0d5b5652",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "archive_table = pd.read_csv(archive_filename)\n",
    "\n",
    "# Remove GRCh37 from the archive list\n",
    "archive_table = archive_table[archive_table[\"version\"] != \"GRCh37\"].reset_index()\n",
    "\n",
    "archive_table[\"numeric_version\"] = archive_table[\"version\"].astype(int)\n",
    "\n",
    "\n",
    "def closest_release(release, archive_table):\n",
    "    if release in archive_table:\n",
    "        return release\n",
    "\n",
    "    return max([V for V in archive_table[\"numeric_version\"] if V <= release])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "337b2890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "closest_release\n",
       "80       985\n",
       "98         9\n",
       "99         5\n",
       "100       21\n",
       "101        8\n",
       "102       16\n",
       "103       12\n",
       "104       17\n",
       "105       10\n",
       "106       35\n",
       "107       12\n",
       "108        4\n",
       "109        4\n",
       "110       11\n",
       "111       52\n",
       "112      354\n",
       "113    34303\n",
       "dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[\"closest_release\"] = 0\n",
    "\n",
    "releases = versions[\"release\"].drop_duplicates().astype(int)\n",
    "\n",
    "# Only have to call closest_release once per version, instead of >70k times\n",
    "for release in releases:\n",
    "    versions.loc[versions[\"release\"] == str(release), \"closest_release\"] = (\n",
    "        closest_release(release, archive_table)\n",
    "    )\n",
    "\n",
    "versions.groupby(\"closest_release\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "343e5006",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000151650.8</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000151650</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000168268.11</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000168268</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000186310.10</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000186310</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000204616.11</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000204616</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000158467.17</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000158467</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly peptide possible_replacement release              latest  type  \\\n",
       "0   GRCh38    None                   []     113   ENSG00000151650.8  Gene   \n",
       "1   GRCh38    None                   []     113  ENSG00000168268.11  Gene   \n",
       "2   GRCh38    None                   []     113  ENSG00000186310.10  Gene   \n",
       "3   GRCh38    None                   []     113  ENSG00000204616.11  Gene   \n",
       "4   GRCh38    None                   []     113  ENSG00000158467.17  Gene   \n",
       "\n",
       "                id  version is_current  closest_release  \\\n",
       "0  ENSG00000151650        8          1              113   \n",
       "1  ENSG00000168268       11          1              113   \n",
       "2  ENSG00000186310       10          1              113   \n",
       "3  ENSG00000204616       11          1              113   \n",
       "4  ENSG00000158467       17          1              113   \n",
       "\n",
       "                                           permalink  \n",
       "0  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "1  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "2  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "3  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "4  https://oct2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[\"permalink\"] = \"\"\n",
    "\n",
    "for i in versions.index:\n",
    "    match = archive_table[\"numeric_version\"] == versions.at[i, \"closest_release\"]\n",
    "    url = archive_table.loc[match, \"url\"].to_string(index=False)\n",
    "    if len(url) > 0:\n",
    "        versions.at[i, \"permalink\"] = (\n",
    "            url + \"/Homo_sapiens/Gene/Summary?db=core;g=\" + versions.at[i, \"id\"]\n",
    "        )\n",
    "\n",
    "versions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4b01719d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>peptide</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>release</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th>is_current</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000265108.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000265108</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>ENSG00000280803.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000280803</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000281672.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000281672</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>87</td>\n",
       "      <td>ENSG00000279857.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000279857</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>84</td>\n",
       "      <td>ENSG00000274483.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000274483</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    assembly peptide possible_replacement release             latest  type  \\\n",
       "67    GRCh38    None                   []      84  ENSG00000265108.1  Gene   \n",
       "68    GRCh38    None                   []      80  ENSG00000280803.1  Gene   \n",
       "111   GRCh38    None                   []      84  ENSG00000281672.1  Gene   \n",
       "135   GRCh38    None                   []      87  ENSG00000279857.1  Gene   \n",
       "141   GRCh38    None                   []      84  ENSG00000274483.1  Gene   \n",
       "\n",
       "                  id  version is_current  closest_release  \\\n",
       "67   ENSG00000265108        1                          80   \n",
       "68   ENSG00000280803        1                          80   \n",
       "111  ENSG00000281672        1                          80   \n",
       "135  ENSG00000279857        1                          80   \n",
       "141  ENSG00000274483        1                          80   \n",
       "\n",
       "                                             permalink  \n",
       "67   https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "68   https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "111  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "135  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "141  https://may2015.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[versions[\"closest_release\"] < 100].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c4128cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://oct2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000151650\n",
      "https://oct2024.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000142192\n"
     ]
    }
   ],
   "source": [
    "print(versions[\"permalink\"][0])\n",
    "print(versions[\"permalink\"][25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "73791e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Does every gene have an associated URL?\n",
    "url_base_len = len(archive_table[\"url\"][0]) + 1\n",
    "all([len(url) > url_base_len for url in versions[\"permalink\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b98b62a",
   "metadata": {},
   "source": [
    "# Part 4: Add permalinks to the gene table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f3edfd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35858, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>alias</th>\n",
       "      <th>name</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>notfound</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000151650</td>\n",
       "      <td>27287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[NA88A, HPX42B, VENTX2]</td>\n",
       "      <td>VENT homeobox</td>\n",
       "      <td>This gene encodes a member of the Vent family ...</td>\n",
       "      <td>VENTX</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000168268</td>\n",
       "      <td>64943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>5'-nucleotidase domain containing 2</td>\n",
       "      <td>Predicted to enable 5'-nucleotidase activity. ...</td>\n",
       "      <td>NT5DC2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186310</td>\n",
       "      <td>4675</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MB20, NPL3]</td>\n",
       "      <td>nucleosome assembly protein 1 like 3</td>\n",
       "      <td>This gene is intronless and encodes a member o...</td>\n",
       "      <td>NAP1L3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000204616</td>\n",
       "      <td>11074</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[C6orf13, RNF, HCGI, HCG1]</td>\n",
       "      <td>tripartite motif containing 31</td>\n",
       "      <td>This gene encodes a protein that functions as ...</td>\n",
       "      <td>TRIM31</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000158467</td>\n",
       "      <td>23382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[IRBIT2, ADOHCYASE3]</td>\n",
       "      <td>adenosylhomocysteinase like 2</td>\n",
       "      <td>The protein encoded by this gene acts as a hom...</td>\n",
       "      <td>AHCYL2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id    _id _version                       alias  \\\n",
       "0  ENSG00000151650  27287      1.0     [NA88A, HPX42B, VENTX2]   \n",
       "1  ENSG00000168268  64943      1.0                          []   \n",
       "2  ENSG00000186310   4675      1.0                [MB20, NPL3]   \n",
       "3  ENSG00000204616  11074      1.0  [C6orf13, RNF, HCGI, HCG1]   \n",
       "4  ENSG00000158467  23382      1.0        [IRBIT2, ADOHCYASE3]   \n",
       "\n",
       "                                   name  \\\n",
       "0                         VENT homeobox   \n",
       "1   5'-nucleotidase domain containing 2   \n",
       "2  nucleosome assembly protein 1 like 3   \n",
       "3        tripartite motif containing 31   \n",
       "4         adenosylhomocysteinase like 2   \n",
       "\n",
       "                                             summary  symbol    type_of_gene  \\\n",
       "0  This gene encodes a member of the Vent family ...   VENTX  protein-coding   \n",
       "1  Predicted to enable 5'-nucleotidase activity. ...  NT5DC2  protein-coding   \n",
       "2  This gene is intronless and encodes a member o...  NAP1L3  protein-coding   \n",
       "3  This gene encodes a protein that functions as ...  TRIM31  protein-coding   \n",
       "4  The protein encoded by this gene acts as a hom...  AHCYL2  protein-coding   \n",
       "\n",
       "  notfound ensembl_release possible_replacement  \\\n",
       "0      NaN             113                   []   \n",
       "1      NaN             113                   []   \n",
       "2      NaN             113                   []   \n",
       "3      NaN             113                   []   \n",
       "4      NaN             113                   []   \n",
       "\n",
       "                                           permalink  \n",
       "0  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "1  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "2  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "3  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "4  https://oct2024.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions = versions[[\"id\", \"release\", \"possible_replacement\", \"permalink\"]]\n",
    "versions.rename(\n",
    "    columns={\"id\": \"ensembl_gene_id\", \"release\": \"ensembl_release\"}, inplace=True\n",
    ")\n",
    "\n",
    "gene_table_merged = pd.merge(\n",
    "    left=gene_table_merged,\n",
    "    right=versions,\n",
    "    how=\"left\",\n",
    "    on=\"ensembl_gene_id\",\n",
    "    validate=\"one_to_one\",\n",
    ")\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01eba7c2",
   "metadata": {},
   "source": [
    "### Final cleanup\n",
    "Unfilled \"possible_replacement\" entries should be changed from NaN to empty lists. \n",
    "\n",
    "\"possible_replacement\" entries that have data in them exist as a list of dicts, and need to have the Ensembl IDs pulled out of them as a list of strings. \n",
    "\n",
    "Remove unneeded columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d0c07b7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000151650</td>\n",
       "      <td>VENT homeobox</td>\n",
       "      <td>[NA88A, HPX42B, VENTX2]</td>\n",
       "      <td>This gene encodes a member of the Vent family ...</td>\n",
       "      <td>VENTX</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000168268</td>\n",
       "      <td>5'-nucleotidase domain containing 2</td>\n",
       "      <td>[]</td>\n",
       "      <td>Predicted to enable 5'-nucleotidase activity. ...</td>\n",
       "      <td>NT5DC2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186310</td>\n",
       "      <td>nucleosome assembly protein 1 like 3</td>\n",
       "      <td>[MB20, NPL3]</td>\n",
       "      <td>This gene is intronless and encodes a member o...</td>\n",
       "      <td>NAP1L3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000204616</td>\n",
       "      <td>tripartite motif containing 31</td>\n",
       "      <td>[C6orf13, RNF, HCGI, HCG1]</td>\n",
       "      <td>This gene encodes a protein that functions as ...</td>\n",
       "      <td>TRIM31</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000158467</td>\n",
       "      <td>adenosylhomocysteinase like 2</td>\n",
       "      <td>[IRBIT2, ADOHCYASE3]</td>\n",
       "      <td>The protein encoded by this gene acts as a hom...</td>\n",
       "      <td>AHCYL2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35853</th>\n",
       "      <td>ENSG00000241472</td>\n",
       "      <td>PTPRG antisense RNA 1</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTPRG-AS1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35854</th>\n",
       "      <td>ENSG00000133106</td>\n",
       "      <td>epithelial stromal interaction 1</td>\n",
       "      <td>[BRESI1]</td>\n",
       "      <td>The protein encoded by this gene has been show...</td>\n",
       "      <td>EPSTI1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35855</th>\n",
       "      <td>ENSG00000230373</td>\n",
       "      <td>golgin A6 family like 3, pseudogene</td>\n",
       "      <td>[GOLGA6L3, GOLGA6L21P, GOLGA6L17P]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOLGA6L3P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35856</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>[LOC105377683]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35857</th>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>killer cell immunoglobulin like receptor, two ...</td>\n",
       "      <td>[CD158A, NKAT1, KIR2DL3, LOC124900571, KIR-K64...</td>\n",
       "      <td>Killer cell immunoglobulin-like receptors (KIR...</td>\n",
       "      <td>KIR2DL1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>113</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://oct2024.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35858 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id                                               name  \\\n",
       "0      ENSG00000151650                                      VENT homeobox   \n",
       "1      ENSG00000168268                5'-nucleotidase domain containing 2   \n",
       "2      ENSG00000186310               nucleosome assembly protein 1 like 3   \n",
       "3      ENSG00000204616                     tripartite motif containing 31   \n",
       "4      ENSG00000158467                      adenosylhomocysteinase like 2   \n",
       "...                ...                                                ...   \n",
       "35853  ENSG00000241472                              PTPRG antisense RNA 1   \n",
       "35854  ENSG00000133106                   epithelial stromal interaction 1   \n",
       "35855  ENSG00000230373                golgin A6 family like 3, pseudogene   \n",
       "35856  ENSG00000249738                          uncharacterized LOC285626   \n",
       "35857  ENSG00000276387  killer cell immunoglobulin like receptor, two ...   \n",
       "\n",
       "                                                   alias  \\\n",
       "0                                [NA88A, HPX42B, VENTX2]   \n",
       "1                                                     []   \n",
       "2                                           [MB20, NPL3]   \n",
       "3                             [C6orf13, RNF, HCGI, HCG1]   \n",
       "4                                   [IRBIT2, ADOHCYASE3]   \n",
       "...                                                  ...   \n",
       "35853                                                 []   \n",
       "35854                                           [BRESI1]   \n",
       "35855                 [GOLGA6L3, GOLGA6L21P, GOLGA6L17P]   \n",
       "35856                                     [LOC105377683]   \n",
       "35857  [CD158A, NKAT1, KIR2DL3, LOC124900571, KIR-K64...   \n",
       "\n",
       "                                                 summary     symbol  \\\n",
       "0      This gene encodes a member of the Vent family ...      VENTX   \n",
       "1      Predicted to enable 5'-nucleotidase activity. ...     NT5DC2   \n",
       "2      This gene is intronless and encodes a member o...     NAP1L3   \n",
       "3      This gene encodes a protein that functions as ...     TRIM31   \n",
       "4      The protein encoded by this gene acts as a hom...     AHCYL2   \n",
       "...                                                  ...        ...   \n",
       "35853                                                NaN  PTPRG-AS1   \n",
       "35854  The protein encoded by this gene has been show...     EPSTI1   \n",
       "35855                                                NaN  GOLGA6L3P   \n",
       "35856                                                NaN  LOC285626   \n",
       "35857  Killer cell immunoglobulin-like receptors (KIR...    KIR2DL1   \n",
       "\n",
       "         type_of_gene ensembl_release possible_replacement  \\\n",
       "0      protein-coding             113                   []   \n",
       "1      protein-coding             113                   []   \n",
       "2      protein-coding             113                   []   \n",
       "3      protein-coding             113                   []   \n",
       "4      protein-coding             113                   []   \n",
       "...               ...             ...                  ...   \n",
       "35853           ncRNA             113                   []   \n",
       "35854  protein-coding             113                   []   \n",
       "35855          pseudo             113                   []   \n",
       "35856           ncRNA             113                   []   \n",
       "35857  protein-coding             113                   []   \n",
       "\n",
       "                                               permalink  \n",
       "0      https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "1      https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "2      https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "3      https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "4      https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "...                                                  ...  \n",
       "35853  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "35854  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "35855  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "35856  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "35857  https://oct2024.archive.ensembl.org/Homo_sapie...  \n",
       "\n",
       "[35858 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table_merged[\"possible_replacement\"] = gene_table_merged[\n",
    "    \"possible_replacement\"\n",
    "].apply(lambda cell: cell if cell is not np.NaN else [])\n",
    "\n",
    "gene_table_merged[\"possible_replacement\"] = gene_table_merged.apply(\n",
    "    lambda row: (\n",
    "        row[\"possible_replacement\"]\n",
    "        if len(row[\"possible_replacement\"]) == 0\n",
    "        else [x[\"stable_id\"] for x in row[\"possible_replacement\"]]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "gene_table_merged = gene_table_merged[\n",
    "    [\n",
    "        \"ensembl_gene_id\",\n",
    "        \"name\",\n",
    "        \"alias\",\n",
    "        \"summary\",\n",
    "        \"symbol\",\n",
    "        \"type_of_gene\",\n",
    "        \"ensembl_release\",\n",
    "        \"possible_replacement\",\n",
    "        \"permalink\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "gene_table_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b371435",
   "metadata": {},
   "source": [
    "### Write to a file\n",
    "This will get uploaded to Synapse as [syn25953363](https://www.synapse.org/#!Synapse:syn25953363)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f2287922",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_merged = gene_table_merged.sort_values(by=\"ensembl_gene_id\").reset_index(\n",
    "    drop=True\n",
    ")\n",
    "gene_table_merged\n",
    "gene_table_merged.to_feather(\"../../output/gene_table_merged_GRCh38.p14.feather\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "agora-data-tools-ywFp1Gf9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
